{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc584613",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "import pandas as pd\n",
    "import mappy as mp\n",
    "import pickle\n",
    "from parse_gene_anno import GeneAnno\n",
    "from interlap import reduce\n",
    "from design_probe_m import * \n",
    "from filter_probe import filter_probes\n",
    "import os\n",
    "from collections import Counter\n",
    "import pickle\n",
    "from math import log\n",
    "import mappy as mp\n",
    "from Bio.SeqUtils import MeltingTemp as mt\n",
    "from parse_gene_anno import GeneAnno\n",
    "from interlap import reduce\n",
    "import sys\n",
    "import pandas as pd\n",
    "\n",
    "# Parameter\n",
    "# probe length (52bp) - is it fixed?\n",
    "# acceptable GC content (45-65%), \n",
    "# minimum melting temperature (Tm) (50Â°C), \n",
    "# and other quality control thresholds.\n",
    "\n",
    "# Probe Request: It reads a CSV file (Probes_request.csv) containing the list of genes (by gene symbol) for which probes are needed. It automatically finds the corresponding Ensembl gene IDs.\n",
    "\n",
    "\n",
    "# parameters\n",
    "data_folder = \"probes/saturate_design_genes/\" # 24_02_19_GENES\n",
    "max_exon_probe = 1000000 # the maximum number of probes per exon\n",
    "min_exon_probe = 10 # the minimum number of probes per exon, based on Kai's requirement\n",
    "\n",
    "\n",
    "probe_len = 52\n",
    "min_CDS_probes = 1000000000 #100\n",
    "GC_range = [0.45,0.65] #[0.45,0.65]\n",
    "min_Tm = 50 #50\n",
    "max_Tm_diff = 8 #10\n",
    "num_rep = 5 #5\n",
    "min_ent = 0.9 #0.9\n",
    "\n",
    "best_n = 24 #4\n",
    "min_chain_score = 8 #16\n",
    "\n",
    "save_top = 10\n",
    "\n",
    "\n",
    "min_dist = 50 #50 #40\n",
    "prime_5end = int(probe_len/2)\n",
    "const_3 = 'TAGAAGAGTCTTCCTTTACG'\n",
    "const_5_1 = '/5Phos/GGAGGGCAGCAAACGGAA'\n",
    "const_5_2 = 'NNNNNNNNNNCGAGACTACTGCGAGTAC'\n",
    "\n",
    "\n",
    "# read reference file\n",
    "# human genome sequence,\n",
    "# gene annotation file\n",
    "# prebuilt index of the genome\n",
    "\n",
    "\n",
    "data = pd.read_excel(\"reference/GTEx_Analysis__exon_reads_rank_normalied.xlsx\")\n",
    "data = data[['start','end','gene_id','gene_name','exon_rank','exon_id']]\n",
    "data['gene_id'] = data['gene_id'].str.split('.').str[0]\n",
    "fa_dict = pickle.load(open(\"reference/fa_dict_human.pkl\",\"rb\"))\n",
    "genome_paths = get_reference_paths()\n",
    "gtf_f = genome_paths['gtf_f']\n",
    "index_f = genome_paths['index_f']\n",
    "\n",
    "# load data\n",
    "probe_request = pd.read_csv(data_folder+'Probes_request.csv')\n",
    "probe_request.sort_values(by='gene_id', ascending=False, na_position='last', inplace=True)\n",
    "probe_request.drop_duplicates(subset='gene symbol', keep='first', inplace=True)\n",
    "probe_request.reset_index(drop=True, inplace=True)\n",
    "probe_request = probe_request.applymap(lambda x: x.strip() if isinstance(x, str) else x)\n",
    "missing_genes = update_probe_request_gene_ids(probe_request, data, data_folder)\n",
    "for gene_symbol in missing_genes:\n",
    "    print(gene_symbol)\n",
    "\n",
    "# gene_symbol_map = {\n",
    "#     \"IL17\": \"IL17A\",\n",
    "#     \"IL12\": \"IL12B\",\n",
    "#     \"IL23\": \"IL23A\"\n",
    "# }\n",
    "# for old_symbol, new_symbol in gene_symbol_map.items():\n",
    "#     probe_request.loc[probe_request['gene symbol'] == old_symbol, 'gene symbol'] = new_symbol\n",
    "# missing_genes = update_probe_request_gene_ids(probe_request, data, data_folder)\n",
    "# for gene_symbol in missing_genes:\n",
    "#     print(gene_symbol)\n",
    "# probe_request = pd.read_csv(data_folder+'Updated_probe_request.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258ee1eb",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "aln = mp.Aligner(fn_idx_in=index_f, preset=\"sr\", best_n=best_n, min_chain_score=min_chain_score)\n",
    "an = GeneAnno(gtf_f)\n",
    "\n",
    "# Generating candidate probes\n",
    "# then filter the probes with the qc function _get_probe_from_intervals\n",
    "status_cnt, pickle_probe = get_exon_probe_list(an, aln, fa_dict, probe_request['gene_id'].tolist(), probe_len, min_CDS_probes, GC_range, min_Tm, max_Tm_diff, num_rep, min_ent) \n",
    "pickle.dump(pickle_probe, open(data_folder+\"pickle_probe.pkl\",\"wb\"))\n",
    "probe_dict = pickle.load(open(data_folder+\"pickle_probe.pkl\",\"rb\"))\n",
    "\n",
    "\n",
    "# select the best probes. \n",
    "# based on:\n",
    "# Spacing: It ensures that selected probes for a single gene are not too close to each other, enforcing a min_dist between them.\n",
    "\n",
    "# Prioritization: It may prioritize probes that fall within the protein-coding sequence (CDS) of an exon.\n",
    "\n",
    "# Best of the Best: If there are still too many good probes, it sorts them by how close their GC content is to an ideal value (e.g., 55%) and picks the top candidates.\n",
    "\n",
    "# Final Count: It ensures each gene ends up with a specific number of probes (e.g., between min_exon_probe and max_exon_probe).\n",
    "\n",
    "\n",
    "selected_probe = filter_probes(probe_dict, max_exon_probe, min_exon_probe, min_dist)\n",
    "selected_probe_exon = get_matching_exons(selected_probe, data)\n",
    "\n",
    "\n",
    "# Finally, the save_design_probes function prepares the output files.\n",
    "\n",
    "# Probe Assembly: For each selected 52 bp homology sequence, it attaches constant adapter sequences (const_3 and const_5_1/const_5_2) to create the final, full-length probe sequences that will be synthesized. This step creates the two separate oligos (probe_3 and probe_5) needed for the assay.\n",
    "\n",
    "# Saving Results: It writes the results to CSV files:\n",
    "\n",
    "# Probe_exon_design_all.csv: Contains all the successfully designed probes.\n",
    "\n",
    "# Probe_exon_design_top.csv: A smaller, curated list containing the top save_top (e.g., 10) probes for each gene, ready for ordering.\n",
    "\n",
    "save_design_probes(data_folder, probe_request, selected_probe_exon, const_3, const_5_1, const_5_2, save_top=save_top, filter_same_exon=False)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
