{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d317473",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This script takes the Hypr-seq adata and the loom file as input\n",
    "# And output an scRNA-seq data with detailed annotations\n",
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import argparse\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import loompy\n",
    "import logging\n",
    "import configparser\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "import os\n",
    "import argparse\n",
    "import copy\n",
    "import h5py\n",
    "\n",
    "# Configure logging\n",
    "logging.basicConfig(stream=sys.stdout, level=logging.INFO, format='%(message)s')\n",
    "logger = logging.getLogger(__name__)\n",
    "\n",
    "# Read and filter the Hypr-seq anndata\n",
    "def filter_hypr(adata, probe_level=False):\n",
    "    # Remove cells with less than 1000 UMIs per cell\n",
    "    sc.pp.filter_cells(adata, min_counts=1000)\n",
    "    # merge the probe name, making the cell by probe matrix to cell by gene matrix\n",
    "    # Splitting at the underscore to separate gene names from probe numbers\n",
    "    if not probe_level:\n",
    "        gene_names = [name.split('_')[0] for name in adata.var_names]\n",
    "    else:\n",
    "        gene_names = [name for name in adata.var_names]\n",
    "\n",
    "    # Add the gene names as a new column in the DataFrame of the .var slot\n",
    "    adata.var['gene_name'] = gene_names\n",
    "    # Convert the AnnData to a DataFrame for easier manipulation\n",
    "    adata_df = pd.DataFrame(adata.X.T, index=adata.var_names, columns=adata.obs_names)\n",
    "\n",
    "    # Use the gene names to sum the counts\n",
    "    # Group by the new gene names and sum across columns (probes for the same gene)\n",
    "    aggregated_data = adata_df.groupby(adata.var['gene_name']).sum()\n",
    "\n",
    "    # Transpose back to original shape (samples as rows, genes as columns)\n",
    "    aggregated_data = aggregated_data.T\n",
    "\n",
    "    # Create new AnnData object with the aggregated data\n",
    "    adata_aggregated = sc.AnnData(X=aggregated_data)\n",
    "    # Copy the metadata from the original AnnData object\n",
    "    adata_aggregated.obs = adata.obs.copy()\n",
    "    # Optionally, copy over any relevant .uns data (unsupervised annotations, such as PCA, neighbors, etc.)\n",
    "    #adata_aggregated.uns = adata.uns.copy()\n",
    "\n",
    "    return adata_aggregated\n",
    "\n",
    "\n",
    "# function reads the loomfile downloaded from Tapestri portal\n",
    "def read_tapestri_loom(filename):\n",
    "    \"\"\"\n",
    "    Read data from MissionBio's formatted loom file.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    filename : str\n",
    "        Path to the loom file (.loom)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    anndata.AnnData\n",
    "        An anndata object with the following layers:\n",
    "        adata.X: GATK calls\n",
    "        adata.layers['e']: reads with evidence of mutation\n",
    "        adata.layers['no_e']: reads without evidence of mutation\n",
    "    \"\"\"\n",
    "    loom_file = loompy.connect(filename)\n",
    "\n",
    "    variant_names, amplicon_names, chromosome, location = (\n",
    "        loom_file.ra['id'], loom_file.ra['amplicon'], loom_file.ra['CHROM'], loom_file.ra['POS'])\n",
    "\n",
    "    barcodes = [barcode.split('-')[0] for barcode in loom_file.ca['barcode']]\n",
    "    adata = anndata.AnnData(np.transpose(loom_file[:, :]), dtype=loom_file[:, :].dtype)\n",
    "    adata.layers['e'] = np.transpose(loom_file.layers['AD'][:, :])\n",
    "    adata.layers['no_e'] = np.transpose(loom_file.layers['RO'][:, :])\n",
    "    adata.var_names = variant_names\n",
    "    adata.obs_names = barcodes\n",
    "    adata.varm['amplicon'] = amplicon_names\n",
    "    adata.varm['chrom'] = chromosome\n",
    "    adata.varm['loc'] = location\n",
    "\n",
    "    loom_file.close()\n",
    "\n",
    "    return adata\n",
    "\n",
    "# Optional function, visualize the barcode distribution\n",
    "def barcode_rank_plot(adata, minimum=0, xmax=None):\n",
    "\n",
    "    # Sum the UMIs for each cell\n",
    "    cell_umi_counts_all = adata.X.sum(axis=1)\n",
    "    cell_idxs = np.argwhere(cell_umi_counts_all > minimum)\n",
    "    cell_umi_counts = cell_umi_counts_all[cell_idxs]\n",
    "\n",
    "    # Convert to numpy array if it's not already\n",
    "    cell_umi_counts = np.array(cell_umi_counts).flatten()\n",
    "\n",
    "    # Sort the UMI counts in descending order for the rank plot\n",
    "    sorted_umi_counts = np.sort(cell_umi_counts)[::-1]\n",
    "\n",
    "    plt.figure(figsize=(5, 4), dpi=150)\n",
    "    sns.lineplot(x=range(1, len(sorted_umi_counts) + 1), y=sorted_umi_counts)\n",
    "    plt.xlabel('Barcode Rank')\n",
    "    plt.ylabel('UMI Count')\n",
    "    plt.title('Cell Barcode Rank Plot')\n",
    "    plt.yscale('log')  # Log scale for better visualization\n",
    "    plt.xscale('log') # Log scale\n",
    "    if xmax is not None:\n",
    "        plt.xlim(0, xmax)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "# Function for finding the intersecting barcodes between the modalities\n",
    "# We can drop the idea of mudata for simplity\n",
    "def find_intersecting_and_filter(adata_hypr, adata_loom):\n",
    "    \"\"\"\n",
    "    Find the intersecting barcode, reorder, and return new AnnData objects for both datasets.\n",
    "    \"\"\"\n",
    "    # Find intersecting barcodes\n",
    "    obs_1, obs_2 = adata_hypr.obs_names, adata_loom.obs_names\n",
    "    cmn_barcodes, idx_1, idx_2 = np.intersect1d(obs_1, obs_2, return_indices=True)\n",
    "\n",
    "    # Logging the number of barcodes\n",
    "    logger.info(f\"Found {len(obs_1)} barcodes in modality hypr seq\")\n",
    "    logger.info(f\"Found {len(obs_2)} barcodes in modality loom file\")\n",
    "    logger.info(f\"Found {len(idx_1)} intersecting barcodes\")\n",
    "\n",
    "    # Subset and reorder both datasets based on the intersecting indices\n",
    "    adata_hypr_new = adata_hypr[idx_1, :]\n",
    "    adata_loom_new = adata_loom[idx_2, :]\n",
    "\n",
    "    # # Ensure the order of barcodes is the same in both datasets\n",
    "    # adata_hypr_new = adata_hypr_new[cmn_barcodes, :]\n",
    "    # adata_loom_new = adata_loom_new[cmn_barcodes, :]\n",
    "\n",
    "    return adata_hypr_new, adata_loom_new\n",
    "    \n",
    "\n",
    "# Step 1, determine the germline mutation from the loom file\n",
    "# We envision that if certain mutation appeared too many times in the dataset\n",
    "# (e.g., more than 25% cells have the homo mutation, we call it homo germline)\n",
    "def call_germline_mutations(adata_loom, cutoff=0.25):\n",
    "\n",
    "    fraction_het = np.mean(adata_loom.X == 1, axis=0)\n",
    "    het_germline = adata_loom.var_names[fraction_het > cutoff]\n",
    "\n",
    "    fraction_hom = np.mean(adata_loom.X == 2, axis=0)\n",
    "    hom_germline = adata_loom.var_names[fraction_hom > cutoff]\n",
    "\n",
    "    return list(het_germline) + list(hom_germline)\n",
    "\n",
    "\n",
    "# For AAV control, we may ignore mixed cells and bystander effect\n",
    "def call_AAV_control_edits(adata_loom, config_path='AAV_config.ini'):\n",
    "    \"\"\"\n",
    "    Input: Tapestri loom file and mutation configuration \n",
    "    Current we only accept 1 type of AAV control (should be a continuous region)\n",
    "    Output: All the AAV edits found in the loom file. \n",
    "\n",
    "    This function will not modify adata_loom\n",
    "    \"\"\"\n",
    "    # Read configuration settings\n",
    "    config = configparser.ConfigParser()\n",
    "    config.read(config_path)\n",
    "    \n",
    "    chrom_name = config.get('AAV_control', 'chrom_name')\n",
    "    start = config.getint('AAV_control', 'start')\n",
    "    end = config.getint('AAV_control', 'end')\n",
    "    variant_alleles = config.get('AAV_control', 'variant_alleles')\n",
    "    \n",
    "    # Parsing variant names\n",
    "    variant_names = np.asarray(adata_loom.var_names.values)\n",
    "    chrom = [name.split(':')[0] for name in variant_names]\n",
    "    loc = [int(name.split(':')[1]) for name in variant_names]\n",
    "    edit_type = [name.split(':')[2] for name in variant_names]\n",
    "\n",
    "    control_editing = []\n",
    "    complement_rule = {'C': 'G', 'G': 'C', 'A': 'T', 'T': 'A'}\n",
    "    try:\n",
    "        reverse_variant_alleles = complement_rule[variant_alleles[0]] + \"/\" + complement_rule[variant_alleles[-1]]\n",
    "    except:\n",
    "        raise ValueError(f\"The variant alleles is not supported. Current variant alleles is {variant_alleles}\")\n",
    "    # Identify control editing cells\n",
    "    for i in range(len(chrom)):\n",
    "        if chrom[i] == chrom_name and start <= loc[i] < end+1:\n",
    "            if edit_type[i] in [variant_alleles, reverse_variant_alleles]:\n",
    "                control_editing.append(variant_names[i])\n",
    "\n",
    "    return control_editing\n",
    "\n",
    "\n",
    "def call_AAV_del_edits(adata_loom, config_path):\n",
    "    \"\"\"\n",
    "    Input: Tapestri loom file and mutation configuration \n",
    "    Current we only accept 1 type of AAV control (should be a continuous region)\n",
    "    Output: All the AAV edits found in the loom file. \n",
    "    \"\"\"\n",
    "    # Read configuration settings\n",
    "    config = configparser.ConfigParser()\n",
    "    config.read(config_path)\n",
    "    \n",
    "    chrom_name = config.get('AAV_del', 'chrom_name')\n",
    "    start = config.getint('AAV_del', 'start')\n",
    "    end = config.getint('AAV_del', 'end')\n",
    "    \n",
    "    # Parsing variant names\n",
    "    variant_names = np.asarray(adata_loom.var_names.values)\n",
    "    chrom = [name.split(':')[0] for name in variant_names]\n",
    "    loc = [int(name.split(':')[1]) for name in variant_names]\n",
    "    edit_type = [name.split(':')[2] for name in variant_names]\n",
    "\n",
    "    del_editing = []\n",
    "    \n",
    "    # Identify  editing cells\n",
    "    for i in range(len(chrom)):\n",
    "        if chrom[i] == chrom_name and start <= loc[i] < end+1:\n",
    "            if edit_type[i][-1] == \"*\":\n",
    "                del_editing.append(variant_names[i])\n",
    "\n",
    "    return del_editing\n",
    "\n",
    "\n",
    "def call_AAV_cells(adata_hypr, adata_loom, AAV_editing):\n",
    "    \"\"\"\n",
    "    Annotate the hypr adata (scRNA-seq data) by the AAV_edit found in call_AAV_edits.\n",
    "    We do not perform annotation in the function to ensure consistency\n",
    "    \"\"\"\n",
    "    all_idx = []\n",
    "    for edit in AAV_editing:\n",
    "        # get the idx of the edit we found\n",
    "        idx = adata_loom.var_names.get_loc(edit)\n",
    "        # if the value is 1/2 (hete/homo), we annotate the cell as control edit cell\n",
    "        het_idx, hom_idx = [np.flatnonzero(adata_loom.X[:, idx] == i) for i in (1, 2)]\n",
    "        AAV_edit_idx = np.concatenate([het_idx, hom_idx])\n",
    "        AAV_cells_barcode_loom = adata_loom.obs_names[AAV_edit_idx]\n",
    "        # find the AAV cells index in that of the adata_hypr\n",
    "        _, idx1, _ = np.intersect1d(adata_hypr.obs_names, AAV_cells_barcode_loom, return_indices=True)\n",
    "        all_idx.append(idx1)\n",
    "        \n",
    "    return np.unique(np.concatenate(all_idx))\n",
    "\n",
    "\n",
    "\n",
    "def get_nearby_variants(\n",
    "    variant_names, # List of all variant in the loom file\n",
    "    target_loc, # loci of interest\n",
    "    germline_amplicons, # List of Germline mutations \n",
    "    window_size # window size\n",
    "):\n",
    "    \"\"\"\n",
    "        Get nearby variants within a window size of the target_loc.\n",
    "        Args:\n",
    "            variant_names (list): List of variant names.\n",
    "            target_loc (str): Target loci (e.g. chr1:123456:A/G).\n",
    "            germline_amplicons (list): List of germline amplicons.\n",
    "            window_size (int): Window size.\n",
    "        Returns:\n",
    "            list: List of nearby variants.\n",
    "    \"\"\"\n",
    "    chrom_t, loc_t, _ = target_loc.split(\":\")\n",
    "    nearby_variants = []\n",
    "    for variant in variant_names:\n",
    "        chrom, loc, edit_type = variant.split(\":\")\n",
    "        is_valid_edit_type = not edit_type.endswith(\"*\") and re.match(\n",
    "            r\"^[A-Za-z]/[A-Za-z]$\", edit_type\n",
    "        )\n",
    "        is_not_germline = variant not in germline_amplicons\n",
    "        if (\n",
    "            chrom == chrom_t\n",
    "            and (int(loc_t) - window_size) <= int(loc) <= (int(loc_t) + window_size)\n",
    "            and is_valid_edit_type\n",
    "            and is_not_germline\n",
    "        ):\n",
    "            nearby_variants.append(variant)\n",
    "\n",
    "    return nearby_variants\n",
    "\n",
    "\n",
    "\n",
    "def plot_mutant_types(target_loc, new_matrix, mt_type, snp_name, save_path, index=\"\"):\n",
    "    \"\"\"\n",
    "    Plot the nearby genotype heatmap\n",
    "    \"\"\"\n",
    "    mutation_counts = new_matrix.apply(np.count_nonzero, axis=0)\n",
    "    plt.figure(figsize=(10, 6))\n",
    "    bars = plt.bar(mutation_counts.index, mutation_counts.values)\n",
    "    for i, idx in enumerate(mutation_counts.index):\n",
    "        plt.text(\n",
    "            bars[i].get_x() + bars[i].get_width() / 2,\n",
    "            bars[i].get_height(),\n",
    "            str(mutation_counts.values[i]),\n",
    "            ha=\"center\",\n",
    "            va=\"bottom\",\n",
    "        )\n",
    "    if target_loc in mutation_counts.index:\n",
    "        bars[mutation_counts.index.tolist().index(target_loc)].set_color(\"r\")\n",
    "    red_patch = mpatches.Patch(color=\"red\", label=\"Target Loci\")\n",
    "    plt.legend(handles=[red_patch])\n",
    "\n",
    "    # we need to rename the target_loc to ensure we do not introduce extra path problems\n",
    "    target_loc = target_loc.replace(\"/\", \"-\")\n",
    "\n",
    "\n",
    "    plot_title = (\n",
    "        f\"Mutation Counts: {mt_type} - {target_loc} ({index}) with total cells {len(new_matrix)}\"\n",
    "    )\n",
    "    plt.title(plot_title)\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.xlabel(\"\")\n",
    "    plt.ylabel(\"Number of Mutated Cells\")\n",
    "\n",
    "    save_index = f\"{snp_name}_{mt_type}_mutant_counts_{index}.pdf\"\n",
    "    plt.savefig(os.path.join(save_path, save_index), bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "    # Plot a cluster map of matrix but cluster rows only.\n",
    "    sns.clustermap(\n",
    "        new_matrix,\n",
    "        cmap=\"YlGnBu\",\n",
    "        row_cluster=True,\n",
    "        col_cluster=False,\n",
    "        figsize=(10, 10),\n",
    "    )\n",
    "    plt.title(\n",
    "        f\"Mutation Matrix: {mt_type} - {snp_name} aka {target_loc} with shape {new_matrix.shape}\"\n",
    "    )\n",
    "    plt.xlabel(\"Variant\")\n",
    "    plt.ylabel(\"Cell Barcode\")\n",
    "    save_index = f\"{snp_name}_{mt_type}_mutant_matrix_{index}.pdf\"\n",
    "    plt.savefig(os.path.join(save_path, save_index), bbox_inches=\"tight\")\n",
    "    plt.close()\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def process_bystander_cells(target_loc, cells, nearby_variants, adata_loom):\n",
    "    \"\"\"\n",
    "    Identify which cells have bystander editing.\n",
    "    Args:\n",
    "        target_loc (str): Target loci (e.g. chr1:123456:A/G).\n",
    "        cells (list): List of cells (that supposedly have the mutation).\n",
    "        nearby_variant: adata that includes only the nearby cells\n",
    "    Returns:\n",
    "        list: List of pure cells.\n",
    "        pd.DataFrame: New matrix.\n",
    "        pd.DataFrame: Matrix with false amplicons.\n",
    "    \"\"\"\n",
    "    nearby_adata = adata_loom[cells, nearby_variants].copy()\n",
    "    new_matrix = pd.DataFrame(\n",
    "        nearby_adata.X, index=nearby_adata.obs_names, columns=nearby_adata.var_names\n",
    "    )\n",
    "    # Change all 3 to 0 in the new matrix, but excluding the target loci column.\n",
    "    for col in new_matrix.columns:\n",
    "        if col != target_loc:\n",
    "            new_matrix[col] = new_matrix[col].apply(\n",
    "                lambda val: 0 if val == 3 else val\n",
    "            )\n",
    "\n",
    "    # Remove the target loci column to generate a bystander position only matrix.\n",
    "    matrix_without_target_loci = new_matrix.loc[\n",
    "        :, new_matrix.columns != target_loc\n",
    "    ].copy()\n",
    "    bystander_sum_for_every_cell = matrix_without_target_loci.sum(axis=1)\n",
    "    # Remove cells that have a row sum of 0, or there's no bystander editing.\n",
    "    matrix_without_target_loci = matrix_without_target_loci.loc[\n",
    "        bystander_sum_for_every_cell != 0\n",
    "    ]\n",
    "\n",
    "    # If there's no bystander editing, return the original cells.\n",
    "    if len(matrix_without_target_loci) == 0:\n",
    "        # Set the bystander as the empty set\n",
    "        return set(cells), set(), new_matrix\n",
    "\n",
    "    # Everything that's left is a bystander cell.\n",
    "    pure_cells = set(cells) - set(matrix_without_target_loci.index)\n",
    "    return pure_cells, set(matrix_without_target_loci.index), new_matrix\n",
    "\n",
    "\n",
    "\n",
    "def call_single_loci_cells(\n",
    "    target_loc, # the loci of interest\n",
    "    adata_loom, # the loom matrix to retrieve info\n",
    "    window_size, # the window size \n",
    "    germline_amplicons, # The germline mutations\n",
    "    snp_name, # the name of the loci\n",
    "    plot_path, # the path to save the figures\n",
    "):\n",
    "    \"\"\"\n",
    "    Process a single target loci. \n",
    "    We will annotate all cells that include the loci\n",
    "    The values represent the mutant type:\n",
    "        0: Background\n",
    "        1: Pure heterozygous\n",
    "        2: Heterozygous with bystander editing\n",
    "        3: Pure homozygous\n",
    "        4: Homozygous with bystander editing\n",
    "    \"\"\"\n",
    "    # Find the idx of the loci\n",
    "    idx = adata_loom.var_names.get_loc(target_loc)\n",
    "    # Annotate cells that are background, het, and hom based on the loom matrix\n",
    "    bkg_cells, het_cells, hom_cells = [\n",
    "            adata_loom.obs_names[np.flatnonzero(adata_loom.X[:, idx] == i)] for i in (0, 1, 2)\n",
    "            ]\n",
    "    # Find all nearby variant of the give loci\n",
    "    variant_names = adata_loom.var_names\n",
    "    nearby_variants = get_nearby_variants(\n",
    "            variant_names, target_loc, germline_amplicons, window_size\n",
    "    )\n",
    "    if target_loc == \"chr5:134138057:A/G\":\n",
    "        if 'chr5:134138060:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138060:A/G\")\n",
    "        if 'chr5:134138062:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138062:A/G\")\n",
    "\n",
    "    if target_loc == \"chr5:134138060:A/G\":\n",
    "        if 'chr5:134138057:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138060:A/G\")\n",
    "        if 'chr5:134138062:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138062:A/G\")\n",
    "\n",
    "    if target_loc == \"chr5:134138062:A/G\":\n",
    "        if 'chr5:134138057:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138060:A/G\")\n",
    "        if 'chr5:134138060:A/G' in nearby_variants:\n",
    "            nearby_variants.remove(\"chr5:134138062:A/G\")\n",
    "    \n",
    "\n",
    "    het_pure, het_bystander, het_matrix = process_bystander_cells(\n",
    "            target_loc, het_cells, nearby_variants, adata_loom\n",
    "    )\n",
    "    hom_pure, hom_bystander, hom_matrix = process_bystander_cells(\n",
    "            target_loc, hom_cells, nearby_variants, adata_loom\n",
    "    )\n",
    "    plot_mutant_types(target_loc, het_matrix, \"Heterozygous\", snp_name, plot_path, \"1-0\")\n",
    "    plot_mutant_types(target_loc, hom_matrix, \"Homozygous\", snp_name, plot_path, \"1-0\")\n",
    "\n",
    "    return list(het_bystander), list(hom_bystander), list(het_pure), list(hom_pure)\n",
    "\n",
    "    # adata_hypr.loc[hom_pure, \"genotype\"] = f\"{target_loc}_homo_pure\"\n",
    "    # adata_hypr.loc[het_pure, \"genotype\"] = f\"{target_loc}_hete_pure\"\n",
    "\n",
    "\n",
    "def annotate_cells(adata):\n",
    "    \"\"\"\n",
    "    Annotates each cell in adata.obs based on the SNP, control, and del columns, according to specified rules.\n",
    "    \n",
    "    Parameters:\n",
    "    - adata: AnnData object containing the obs DataFrame with binary columns for each condition.\n",
    "    \n",
    "    Returns:\n",
    "    - Annotations are stored in a new column 'annotation' in adata.obs.\n",
    "    \"\"\"\n",
    "    \n",
    "    # List of condition columns, based on the presence of SNPs, control, and del columns in adata.obs\n",
    "    snp_columns = [col for col in adata.obs.columns if any(sub in col for sub in [\"het_bystander\", \"het_pure\", \"hom_bystander\", \"hom_pure\"])]\n",
    "    control_column = \"AAV_control\"\n",
    "    del_column = \"AAV_del\"\n",
    "    \n",
    "    # Define the function for annotating a single cell\n",
    "    def annotate_row(row):\n",
    "        # Check SNP-related rules\n",
    "        snp_active = row[snp_columns].sum()  # Count number of active SNP columns (True values)\n",
    "        \n",
    "        if snp_active > 1:  # More than one SNP column is True\n",
    "            return \"mixed\"\n",
    "        \n",
    "        if row[control_column] and snp_active == 1:  # SNP and control active\n",
    "            # Find which SNP column is active\n",
    "            snp_name = row[snp_columns][row[snp_columns] == True].index[0]\n",
    "            return \"mixed\"\n",
    "        \n",
    "        if row[del_column] and snp_active >= 1:  # SNP and del active\n",
    "            return \"mixed\"\n",
    "        \n",
    "        if row[control_column] and snp_active == 0:  # Only control active\n",
    "            return \"AAV_control\"\n",
    "        \n",
    "        if row[del_column] and snp_active == 0:  # Only del active\n",
    "            return \"AAV_del\"\n",
    "        \n",
    "        if snp_active == 1:  # Single SNP active, no control or del\n",
    "            snp_name = row[snp_columns][row[snp_columns] == True].index[0]\n",
    "            return snp_name\n",
    "        \n",
    "        # If no columns are active\n",
    "        if snp_active == 0 and not row[control_column] and not row[del_column]:\n",
    "            return \"unedited\"\n",
    "        \n",
    "        # Default case, though we should not reach here\n",
    "        return \"unknown\"\n",
    "    # import pdb; pdb.set_trace()\n",
    "    # Apply the annotation function to each row in a vectorized manner\n",
    "    adata.obs['genotype_annotation'] = adata.obs.apply(annotate_row, axis=1)\n",
    "\n",
    "\n",
    "def annotate_genotype(adata_loom, \n",
    "                    adata_hypr, \n",
    "                    config_path,\n",
    "                    save_path=\"./\",\n",
    "                    plot_path=\"./\"):\n",
    "    \"\"\"\n",
    "    Annotate the adata_hypr, such that in the adata_hypr.obs, we have:\n",
    "    Homo pure for SNP 1, 2, 3, …\n",
    "    Homo bystander for SNP 1,2, 3, …\n",
    "    Hete pure for SNP 1, 2, 3, …\n",
    "    Hete bystander for SNP 1, 2, 3…\n",
    "    Mixed cells\n",
    "    AAVs: AAV_control, AAV_del\n",
    "    Unedited. \n",
    "    \"\"\"\n",
    "\n",
    "    # Perform the annotation step. We need:\n",
    "    # 1. Determine the germline mutations\n",
    "    # 2. Annotate the AAV_control cells\n",
    "    # 3. Annotate the AAV_del_control cells\n",
    "    # 4. For each SNP we are interested in, find if it has bystander effect\n",
    "    # 5. Determine the bystander mutations\n",
    "    # 6.  Annotate all mixed cells. For example, for 2 SNPs we are interested in, \n",
    "    #  if they are co-edited, we need to count the number. We need to annotate it as mixed.\n",
    "    \n",
    "\n",
    "    # Step 1, determine the germline mutations\n",
    "    germline_amplicons = call_germline_mutations(adata_loom, cutoff=0.25)\n",
    "\n",
    "    # Step 2, annotate the AAV_control cells\n",
    "    # 2.1 Find the control editing locis \n",
    "    AAV_control_editing = call_AAV_control_edits(adata_loom, config_path=config_path)\n",
    "    # 2.2 Add a column to the adata_hypr obs and then perform the annotation only here\n",
    "    adata_hypr.obs[\"AAV_control\"] = False\n",
    "    AAV_control_cell_idx = call_AAV_cells(adata_hypr, adata_loom, AAV_editing=AAV_control_editing)\n",
    "    adata_hypr.obs.iloc[AAV_control_cell_idx, adata_hypr.obs.columns.get_loc(\"AAV_control\")] = True\n",
    "\n",
    "    # Step 3, annotate the AAV_del_control_cells\n",
    "    # 3.1 Find the del editing locis\n",
    "    AAV_del_editing = call_AAV_del_edits(adata_loom, config_path=config_path)\n",
    "    # 3.2 Add a column to the adata_hypr and then perform the annotation only here\n",
    "    adata_hypr.obs[\"AAV_del\"] = False\n",
    "    AAV_del_cell_idx = call_AAV_cells(adata_hypr, adata_loom, AAV_editing=AAV_del_editing)\n",
    "    adata_hypr.obs.iloc[AAV_del_cell_idx, adata_hypr.obs.columns.get_loc(\"AAV_del\")] = True\n",
    "\n",
    "    # Step 4/5, Enumerate all other variants we are interested in\n",
    "    # Initialize the config parser\n",
    "    config = configparser.ConfigParser()\n",
    "    # Read the config file\n",
    "    config.read(config_path)\n",
    "    # Retrieve and return all section names\n",
    "    sections = config.sections()\n",
    "    exclude_sections = [\"AAV_control\", \"AAV_del\"]\n",
    "    SNPs = [section for section in sections if section not in exclude_sections]\n",
    "\n",
    "    # Each snp is the section name we defined in the config file\n",
    "    # Perform the annotation here\n",
    "    annotation = [\"het_bystander\", \"het_pure\", \"hom_bystander\", \"hom_pure\"]\n",
    "\n",
    "    for snp in SNPs:\n",
    "        chrom_name = config.get(snp, 'chrom_name')\n",
    "        locus = config.getint(snp, 'locus')\n",
    "        alleles = config.get(snp, 'variant_alleles')\n",
    "        complement_rule = {'C': 'G', 'G': 'C', 'A': 'T', 'T': 'A'}\n",
    "        try:\n",
    "            complement_variant_alleles = complement_rule[alleles[0]] + \"/\" + complement_rule[alleles[-1]]\n",
    "        except:\n",
    "            raise ValueError(f\"The variant alleles is not supported. Current variant alleles is {alleles}\")\n",
    "        # Transfer back to the original version\n",
    "        target_loc = \":\".join([chrom_name, str(locus), alleles])\n",
    "        complement_target_loc = \":\".join([chrom_name, str(locus), complement_variant_alleles])\n",
    "        \n",
    "        # Find which allele is the one we want, the original version or the complement one\n",
    "        if target_loc in adata_loom.var_names:\n",
    "            target_loc = target_loc\n",
    "        elif complement_target_loc in adata_loom.var_names:\n",
    "            target_loc = complement_target_loc\n",
    "        else:\n",
    "            raise ValueError(f\"Both the mutation {target_loc} and the reversed one {complement_target_loc} is not found in the loom file\")\n",
    "\n",
    "\n",
    "        # Get the index of the cells\n",
    "        het_bystander, hom_bystander, het_pure, hom_pure = call_single_loci_cells(target_loc, adata_loom, window_size=10, \n",
    "            germline_amplicons=germline_amplicons, snp_name = snp, plot_path=plot_path)\n",
    "        \n",
    "\n",
    "        for i, idx_i in enumerate([het_bystander, het_pure, hom_bystander, hom_pure]):\n",
    "            adata_hypr.obs[f\"{snp}_{annotation[i]}\"] = False\n",
    "            adata_hypr.obs.loc[idx_i, f\"{snp}_{annotation[i]}\"] = True\n",
    "\n",
    "        # adata_hypr.obs[f\"{snp}\"] = adata_hypr.obs[f\"{snp}_het_all\"] | adata_hypr.obs[f\"{snp}_hom_all\"]\n",
    "\n",
    "    # Step 7, for the cells that has genotypes, we need to ensure that they are not mixed cells\n",
    "    annotate_cells(adata_hypr)\n",
    "\n",
    "    return adata_hypr, adata_loom\n",
    "    \n",
    "\n",
    "\n",
    "\n",
    "def main():\n",
    "    # Create the parser\n",
    "    parser = argparse.ArgumentParser(description=\"Filter, Intersect and Annotate \\\n",
    "     the genotype information to the phenotype RNA-seq count matrix\")\n",
    "    \n",
    "    # Add arguments for file paths\n",
    "    parser.add_argument('--loom_path', type=str, required=True, help='The path to the loom file -- genotype')\n",
    "    parser.add_argument('--hypr_path', type=str, required=True, help='The path to the hypr file -- phenotype')\n",
    "    parser.add_argument('--config_path', type=str, required=True, help='The path to the variant configuration file')\n",
    "    parser.add_argument('--save_path', type=str, required=True, help='The path to save annotated anndata')\n",
    "    parser.add_argument('--plot_path', type=str, required=True, help='The path to save genotype clustermap figure')\n",
    "    parser.add_argument('--probe_level', type=int, required=True, help=\"1 for probe level matrix, other numbers for gene_level\")\n",
    "    \n",
    "    # Parse the arguments\n",
    "    args = parser.parse_args()\n",
    "    \n",
    "\n",
    "    # read the data and set the config path\n",
    "    adata_loom = read_tapestri_loom(args.loom_path)\n",
    "    adata_hypr = sc.read(args.hypr_path)\n",
    "    # import pdb; pdb.set_trace()\n",
    "    probe_level = (args.probe_level==1)\n",
    "\n",
    "    adata_hypr = filter_hypr(adata_hypr, probe_level=probe_level)\n",
    "\n",
    "    config_path = args.config_path\n",
    "    save_path = args.save_path\n",
    "    plot_path = args.plot_path\n",
    "\n",
    "    if not os.path.exists(save_path):\n",
    "        # Create the directory, including any intermediate directories\n",
    "        os.makedirs(save_path)\n",
    "    if not os.path.exists(plot_path):\n",
    "        # Create the directory, including any intermediate directories\n",
    "        os.makedirs(plot_path)\n",
    "\n",
    "    # filter and intersect between hypr matrix and loom matrix\n",
    "    adata_hypr, adata_loom = find_intersecting_and_filter(adata_hypr, adata_loom)\n",
    "    \n",
    "    # Annotate the filtered data\n",
    "    adata_hypr, adata_loom = annotate_genotype(adata_loom=adata_loom, adata_hypr=adata_hypr, config_path=config_path, save_path=save_path, plot_path=plot_path)\n",
    "    if probe_level:\n",
    "        hypr_name = f\"{save_path}/Annotated_phenotype_hypr_seq_probe.h5ad\"\n",
    "    else:\n",
    "        hypr_name = f\"{save_path}/Annotated_phenotype_hypr_seq.h5ad\"\n",
    "    sc.write(hypr_name, adata_hypr)\n",
    "    sc.write(f\"{save_path}/genotype_loom.h5ad\", adata_loom)\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # example usage:\n",
    "    # python merge_and_annotate.py \\\n",
    "    # --loom_path=\"data/TCF7_CD4_8/TCF7_CD4-8.cells.loom\" \\\n",
    "    # --hypr_path=\"data/TCF7_CD4_8/TCF7_CD4-8_hypr.h5ad\" \\\n",
    "    # --config_path=\"TCF7.ini\" --save_path=\"result/TCF7_CD4_8/\"\n",
    "    # --plot_path=\"analysis_result/genotype_by_loc/CD4_8\"\n",
    "    main() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3baf89",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Functions For conducting the DEG analysis and plot figures\n",
    "# They should be generalizable for any datasets that generated through the \n",
    "# \"merge_and_annotate.py\"\n",
    "\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import numpy as np \n",
    "import pandas as pd \n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "def compute_marker_genes_and_plot(adata, condition_1, condition_2, task_name=\"\"):\n",
    "    \"\"\"\n",
    "    We need to ensure that the condition_2 is the control case. Otherwise, the \n",
    "    saved parameters/values may not so reliable.\n",
    "    \"\"\"\n",
    "    \n",
    "    # assert condition_2 in [\"AAV_control\", \"unedited\", \"AAV_del\"]\n",
    "    # import pdb; pdb.set_trace()\n",
    "    filtered_adata = adata[(adata.obs[\"genotype_annotation\"]==condition_1) | (adata.obs[\"genotype_annotation\"]==condition_2)]\n",
    "\n",
    "    filtered_adata.obs['genotype_annotation'] = filtered_adata.obs['genotype_annotation'].astype('category')\n",
    "    # import pdb; pdb.set_trace()\n",
    "    sc.tl.rank_genes_groups(filtered_adata, \"genotype_annotation\", method=\"wilcoxon\")\n",
    "    # sc.pl.rank_genes_groups(filtered_adata, n_genes=25, sharey=False, save=f\"{cell_type}_{condition_1}_VS_{condition_2}.png\")\n",
    "\n",
    "    # sc.pl.rank_genes_groups_dotplot(filtered_adata, save=f\"{cell_type}_{condition_1}_VS_{condition_2}.png\")\n",
    "    p_val = filtered_adata.uns[\"rank_genes_groups\"][\"pvals_adj\"][condition_1]\n",
    "    log_fc = filtered_adata.uns[\"rank_genes_groups\"][\"logfoldchanges\"][condition_1]\n",
    "    gene_names = filtered_adata.uns[\"rank_genes_groups\"][\"names\"][condition_1]\n",
    "\n",
    "    df = pd.DataFrame({\n",
    "        'gene_name': gene_names,\n",
    "        f'{task_name}_{condition_1}_VS_{condition_2}_log_fc': log_fc,\n",
    "        f'{task_name}_{condition_1}_VS_{condition_2}_p_val': p_val\n",
    "    })\n",
    "    df = df.sort_values(\"gene_name\")\n",
    "    df = df.set_index(\"gene_name\")\n",
    "\n",
    "    return df\n",
    "\n",
    "\n",
    "\n",
    "def plot_genotype_gene_clustermap(adata, genes_of_interest, genotypes_of_interest, task_name=\"\", scale=True):\n",
    "    \"\"\"\n",
    "    Plots a clustermap of average gene expression values for each genotype across a subset of genes.\n",
    "    \n",
    "    Parameters:\n",
    "    - adata: AnnData object containing the single-cell RNA-seq data.\n",
    "    - genes_of_interest: List of gene names (subset of adata.var_names) to include in the plot.\n",
    "    - genotypes_of_interest: List of genotypes (subset of adata.obs[\"genotype_annotation\"]) to include.\n",
    "    - scale: Boolean, if True, scales the expression values for better visualization.\n",
    "    \n",
    "    Returns:\n",
    "    - Clustermap of the average gene expression values for the specified genotypes and genes.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Filter cells by genotypes of interest\n",
    "    adata_subset = adata[adata.obs[\"genotype_annotation\"].isin(genotypes_of_interest), genes_of_interest]\n",
    "    \n",
    "    # Extract expression data and add genotype labels\n",
    "    gene_expression_df = pd.DataFrame(adata_subset.X.toarray(), \n",
    "                                      index=adata_subset.obs_names, \n",
    "                                      columns=adata_subset.var_names)\n",
    "    gene_expression_df[\"genotype\"] = adata_subset.obs[\"genotype_annotation\"].values\n",
    "    \n",
    "    # Compute the average expression per gene for each genotype\n",
    "    avg_expression_df = gene_expression_df.groupby(\"genotype\").mean()\n",
    "    \n",
    "    # Optionally scale the data for visualization\n",
    "    if scale:\n",
    "        avg_expression_df = (avg_expression_df - avg_expression_df.mean()) / avg_expression_df.std()\n",
    "    \n",
    "    # Plot the clustermap\n",
    "    # plt.figure(figsize=(20, 10))\n",
    "    sns.clustermap(avg_expression_df, \n",
    "                   cmap=\"vlag\", \n",
    "                   row_cluster=True, \n",
    "                   col_cluster=True,\n",
    "                   xticklabels=True, \n",
    "                   yticklabels=True,\n",
    "                   figsize=(18, 10))\n",
    "    import os\n",
    "    os.makedirs(f\"analysis_result/genotype_gene_heatmap/{task_name}\",exist_ok=True)\n",
    "    plt.title(f\"Gene Expression Clustermap -- {task_name}\")\n",
    "    plt.tight_layout()\n",
    "    \n",
    "    plt.savefig(f\"analysis_result/genotype_gene_heatmap/{task_name}/heatmap.svg\")\n",
    "    plt.savefig(f\"analysis_result/genotype_gene_heatmap/{task_name}/heatmap.png\")\n",
    "    avg_expression_df.to_csv(f\"analysis_result/genotype_gene_heatmap/{task_name}/heatmap.csv\")\n",
    "\n",
    "\n",
    "\n",
    "def plot_genotype_cluster_composition(adata, genotype_column='genotype_annotation', cluster_column='leiden', save_path=\"./\", task_name=\"\"):\n",
    "    \"\"\"\n",
    "    Plots the composition of leiden clusters within each genotype as a stacked bar chart.\n",
    "    \n",
    "    Parameters:\n",
    "    - adata: AnnData object containing the data.\n",
    "    - genotype_column: str, the column in adata.obs representing genotype annotations.\n",
    "    - cluster_column: str, the column in adata.obs representing leiden cluster labels.\n",
    "    \"\"\"\n",
    "    \n",
    "    # Extract the relevant columns from adata.obs\n",
    "    df = adata.obs[[genotype_column, cluster_column]]\n",
    "    \n",
    "    # Count the occurrences of each leiden cluster within each genotype\n",
    "    composition = df.groupby([genotype_column, cluster_column]).size().unstack(fill_value=0)\n",
    "    \n",
    "    # Normalize counts to get the proportion of each leiden cluster within each genotype\n",
    "    composition = composition.div(composition.sum(axis=1), axis=0)\n",
    "    \n",
    "    # Plot the composition as a stacked bar chart\n",
    "    ax = composition.plot(kind='bar', stacked=True, figsize=(10, 7), colormap=\"tab20\")\n",
    "    \n",
    "    # Add labels and title\n",
    "    plt.title('Leiden Cluster Composition by Genotype')\n",
    "    plt.xlabel('Genotype')\n",
    "    plt.ylabel('Proportion of Each Cluster')\n",
    "    plt.legend(title=cluster_column, bbox_to_anchor=(1.05, 1), loc='upper left')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f\"{save_path}/{task_name}_composition.svg\")\n",
    "    plt.savefig(f\"{save_path}/{task_name}_composition.png\")\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # Here I write simple test for the functions implemented in the file\n",
    "    adata = sc.read(\"/home/hanwenka/project/stag_analysis/result/TCF7_CD4_8/Annotated_phenotype_hypr_seq.h5ad\")\n",
    "\n",
    "    df = compute_marker_genes_and_plot(adata, \"TCF7_SPA6_hom_bystander\", \"AAV_control\", task_name=\"all_cell\")\n",
    "    print(df)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9ded5ffa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# This code should only work for the CD4/8 dataset.\n",
    "\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import numpy as np \n",
    "import pandas as pd \n",
    "\n",
    "from analysis_utils import compute_marker_genes_and_plot, plot_genotype_gene_clustermap, plot_genotype_cluster_composition\n",
    "import seaborn as sns\n",
    "import os\n",
    "from pathlib import Path\n",
    "from scipy.stats import zscore\n",
    "\n",
    "# Do the pairwise deg analysis for the cd4/8 dataset\n",
    "\n",
    "def compute_pairwise_degs_cd4_8(adata):\n",
    "    # The input adata should be annotated with cd4/8 celltypes in\n",
    "    # the adata.obs\n",
    "    assert \"cell_type\" in adata.obs.columns\n",
    "    cd4_cell = adata[adata.obs[\"cell_type\"]==\"CD4\"]\n",
    "    cd8_cell = adata[adata.obs[\"cell_type\"]==\"CD8\"]\n",
    "\n",
    "    all_dfs_cd4 = []\n",
    "    all_dfs_cd8 = []\n",
    "    for cond_1_prefix in ['TCF7_rs12659489', 'TCF7_rs57918439', 'TCF7_SPA4', 'TCF7_SPA6']:\n",
    "        for subfix in [\"het_bystander\", \"het_pure\", \"hom_bystander\", \"hom_pure\"]:\n",
    "            cond1 = f\"{cond_1_prefix}_{subfix}\"\n",
    "            for cond2 in [\"AAV_control\", \"unedited\", \"AAV_del\"]:\n",
    "                result_df_cd4 = compute_marker_genes_and_plot(cd4_cell, condition_1=cond1, condition_2=cond2, task_name=\"CD4\")\n",
    "                result_df_cd8 = compute_marker_genes_and_plot(cd8_cell, condition_1=cond1, condition_2=cond2, task_name=\"CD8\")\n",
    "\n",
    "                all_dfs_cd4.append(result_df_cd4)\n",
    "                all_dfs_cd8.append(result_df_cd8)\n",
    "\n",
    "    final_cd4 = pd.concat(all_dfs_cd4, axis=1)\n",
    "    final_cd8 = pd.concat(all_dfs_cd8, axis=1)\n",
    "\n",
    "    final_cd4.to_csv(\"analysis_result/deg_tables/cd4_genotype_pairwise_deg.csv\")\n",
    "    final_cd8.to_csv(\"analysis_result/deg_tables/cd8_genotype_pairwise_deg.csv\")\n",
    "\n",
    "\n",
    "\n",
    "def compute_leiden_cluster_degs(filtered_adata):\n",
    "    sc.tl.rank_genes_groups(filtered_adata, \"leiden\", method=\"wilcoxon\")\n",
    "\n",
    "    all_dfs = []\n",
    "    for cluster_name in [\"0\", \"1\", \"2\", \"3\", \"4\", \"5\"]:\n",
    "        # import pdb; pdb.set_trace()\n",
    "        p_val = filtered_adata.uns[\"rank_genes_groups\"][\"pvals_adj\"][cluster_name]\n",
    "        log_fc = filtered_adata.uns[\"rank_genes_groups\"][\"logfoldchanges\"][cluster_name]\n",
    "        gene_names = filtered_adata.uns[\"rank_genes_groups\"][\"names\"][cluster_name]\n",
    "\n",
    "        df = pd.DataFrame({\n",
    "            'gene_name': gene_names,\n",
    "            f'{cluster_name}_log_fc': log_fc,\n",
    "            f'{cluster_name}_p_val': p_val\n",
    "        })\n",
    "        df = df.sort_values(\"gene_name\")\n",
    "        df = df.set_index(\"gene_name\")\n",
    "        all_dfs.append(df)\n",
    "    all_dfs = pd.concat(all_dfs, axis=1)\n",
    "\n",
    "    all_dfs.to_csv(\"analysis_result/deg_tables/cd4_8_cluster_degs.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def gene_genotype_clustermap(adata, gene_list_path):\n",
    "    \"\"\"\n",
    "    In this code, I will plot the cluster map figure. One axis is the gene, another axis is the genotype\n",
    "    \"\"\"\n",
    "\n",
    "    # read the gene of interest list\n",
    "    gene_list = pd.read_csv(gene_list_path, header=None)\n",
    "    gene_list.columns = [\"gene\"]\n",
    "    gene_list = gene_list[\"gene\"]\n",
    "\n",
    "    # make the genotype of interest list\n",
    "    genotype_list = [\"AAV_control\"]\n",
    "    for prefix in ['TCF7_rs12659489', 'TCF7_rs57918439', 'TCF7_SPA4', 'TCF7_SPA6']:\n",
    "        for subfix in [\"het_bystander\", \"het_pure\", \"hom_bystander\", \"hom_pure\"]:\n",
    "            genotype = f\"{prefix}_{subfix}\"\n",
    "            genotype_list.append(genotype)\n",
    "    \n",
    "    assert \"cell_type\" in adata.obs.columns\n",
    "    cd4_cell = adata[adata.obs[\"cell_type\"]==\"CD4\"]\n",
    "    cd8_cell = adata[adata.obs[\"cell_type\"]==\"CD8\"]\n",
    "\n",
    "    plot_genotype_gene_clustermap(adata, genes_of_interest=gene_list, genotypes_of_interest=genotype_list, task_name=\"cd4-8_scale\")\n",
    "    # plot_genotype_gene_clustermap(adata, genes_of_interest=gene_list, genotypes_of_interest=genotype_list, task_name=\"no_scale\", scale=False)\n",
    "    plot_genotype_gene_clustermap(cd4_cell, genes_of_interest=gene_list, genotypes_of_interest=genotype_list, task_name=\"cd4_scale\")\n",
    "    plot_genotype_gene_clustermap(cd8_cell, genes_of_interest=gene_list, genotypes_of_interest=genotype_list, task_name=\"cd8_scale\")\n",
    "\n",
    "\n",
    "def plot_umap_figure(adata):\n",
    "    \"\"\"\n",
    "    Plot the umap of the saved adata.\n",
    "    \"\"\"\n",
    "    sc.pl.umap(adata, color=\"leiden\", save=\"_leiden.png\", palette=\"Set1\")\n",
    "    sc.pl.umap(adata, color=[\"CD4\", \"CD8A\", \"CD8B\"], save=\"_cd4_8A_8B.png\", palette=\"Set1\")\n",
    "    sc.pl.umap(adata, color=[\"TCF7\", \"SELL\", \"CD44\"], save=\"_TCF7_SELL_CD44.png\", palette=\"Set1\")\n",
    "    import shutil\n",
    "    shutil.move(\"figures/\",\"CD4_8_umap/\")\n",
    "    shutil.move(\"CD4_8_umap/\", \"analysis_result/umap/\")\n",
    "\n",
    "\n",
    "def plot_composition_figure(adata):\n",
    "    cd4_cell = adata[adata.obs[\"cell_type\"]==\"CD4\"]\n",
    "    cd8_cell = adata[adata.obs[\"cell_type\"]==\"CD8\"]\n",
    "\n",
    "    save_path = \"analysis_result/composition_figures\"\n",
    "    if not os.path.exists(save_path):\n",
    "        # Create the directory, including any intermediate directories\n",
    "        os.makedirs(save_path)\n",
    "    plot_genotype_cluster_composition(cd4_cell, genotype_column=\"genotype_annotation\", cluster_column=\"leiden\", save_path=f\"{save_path}\", task_name=\"cd4\")\n",
    "    plot_genotype_cluster_composition(cd8_cell, genotype_column=\"genotype_annotation\", cluster_column=\"leiden\", save_path=f\"{save_path}\", task_name=\"cd8\")\n",
    "\n",
    "\n",
    "\n",
    "def plot_cluster_by_gene_heatmap(adata):\n",
    "    \"\"\"\n",
    "    This function plot the cluster by gene matrix plot and save the matrix for other plots\n",
    "    \"\"\"\n",
    "    cd8_cell = adata[adata.obs[\"cell_type\"]==\"CD8\"]\n",
    "    Proliferating = [\"RRM2\", \"SLC29A1\", \"CCNB1\", \"CDK2\", \"MYBL2\", \"H2AX\", \"LMNB1\", \"AURKA\"]\n",
    "    naive = [\"PECAM1\", \"CISH\", \"HAVCR2\", \"BACH2\", \"NELL2\", \"TCF7\", \"LEF1\", \"CCR7\", \"SELL\"]\n",
    "    effector = [\"NKG7\", \"KLRD1\", \"GZMH\", \"GZMB\", \"GZMA\", \"CCR5\", \"IFNG\", \"TBX21\", \"TIGIT\", \"BHLHE40\"]\n",
    "\n",
    "    gene_list = Proliferating + naive + effector\n",
    "    \n",
    "    # Make sure all genes are in the dataset\n",
    "    gene_list = [gene for gene in gene_list if gene in adata.var_names]\n",
    "\n",
    "    grouped_means = cd8_cell[:, gene_list].to_df().groupby(cd8_cell.obs['leiden']).mean()\n",
    "\n",
    "    sc._settings.ScanpyConfig.figdir = Path(\"analysis_result/cluster_by_gene_heatmap\")\n",
    "    matrix_result = sc.pl.matrixplot(cd8_cell, var_names=gene_list, groupby=\"leiden\", cmap=\"Blues\", swap_axes=False, dendrogram=False, standard_scale='var', return_fig=True)\n",
    "    # sc.pl.matrixplot(cd8_cell, var_names=gene_list, groupby=\"leiden\", cmap=\"Blues\", swap_axes=False, dendrogram=False, standard_scale='var', save=\"cd8.png\")\n",
    "    \n",
    "    # Save the data\n",
    "    output_dir = Path(\"analysis_result/cluster_by_gene_heatmap\")\n",
    "    output_dir.mkdir(parents=True, exist_ok=True)\n",
    "    scaled_means = grouped_means.copy()\n",
    "    scaled_means -= scaled_means.min(axis=0)  # Subtract the column-wise minimum\n",
    "    scaled_means = (scaled_means / scaled_means.max(axis=0)).fillna(0)  # Divide by the column-wise maximum and handle NaNs\n",
    "\n",
    "    grouped_means.to_csv(output_dir / \"cd8_expression_matrix.csv\")\n",
    "    scaled_means.to_csv(output_dir / \"cd8_expression_matrix_norm.csv\")\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def probe_level_deg():\n",
    "    # the probe cell type is annotated directly using the gene-level clusters\n",
    "    adata = sc.read(\"result/TCF7_CD4_8/Annotated_phenotype_hypr_seq_probe.h5ad\")\n",
    "    # do the following 2 for probes only \n",
    "    sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "    sc.pp.log1p(adata)\n",
    "    adata.obs['genotype_annotation'] = adata.obs[\"genotype_annotation\"].astype('category')\n",
    "    compute_pairwise_degs_cd4_8(adata)\n",
    "\n",
    "\n",
    "def get_umi_count_per_gene_cell_type_cluster():\n",
    "    adata_raw = sc.read(\"/home/hanwenka/project/stag_analysis/result/TCF7_CD4_8/Annotated_phenotype_hypr_seq.h5ad\")\n",
    "    adata_cell_type = sc.read(\"/home/hanwenka/project/stag_analysis/result/TCF7_CD4_8/annotated_genotype_celltype.h5ad\")\n",
    "    adata_raw.obs[\"leiden\"] = adata_cell_type.obs[\"leiden\"]\n",
    "    adata_raw.obs[\"cell_type\"] = adata_cell_type.obs[\"cell_type\"]\n",
    "\n",
    "    grouped_means_leiden = adata_raw.to_df().groupby(adata_raw.obs['leiden']).mean()\n",
    "    grouped_means_cell_type = adata_raw.to_df().groupby(adata_raw.obs['cell_type']).mean()\n",
    "    output_dir = Path(\"analysis_result/mean_umi_clusters_cd_4_8\")\n",
    "    output_dir.mkdir(parents=True, exist_ok=True)\n",
    "    grouped_means_leiden.to_csv(output_dir / \"leiden_expression_matrix.csv\")\n",
    "    grouped_means_cell_type.to_csv(output_dir / \"cell_type_expression_matrix.csv\")\n",
    "    # import pdb; pdb.set_trace()\n",
    "\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    adata = sc.read(\"result/TCF7_CD4_8/annotated_genotype_celltype.h5ad\")\n",
    "    #compute_pairwise_degs_cd4_8(adata)\n",
    "    # compute_leiden_cluster_degs(adata)\n",
    "    # gene_genotype_clustermap(adata, \"/home/hanwenka/project/stag_analysis/internal_files/cd8_heatmap_genes.csv\")\n",
    "    # plot_umap_figure(adata)\n",
    "    # plot_composition_figure(adata)\n",
    "    plot_cluster_by_gene_heatmap(adata)\n",
    "    # get_umi_count_per_gene_cell_type_cluster()\n"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
