{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8677215d",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'mixed_cell_level_qc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mNameError\u001b[39m                                 Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 259\u001b[39m\n\u001b[32m    251\u001b[39m     plt.savefig(\u001b[33m\"\u001b[39m\u001b[33mamplicon_level_quality_box.svg\u001b[39m\u001b[33m\"\u001b[39m, dpi=\u001b[32m300\u001b[39m)\n\u001b[32m    254\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m \u001b[34m__name__\u001b[39m == \u001b[33m\"\u001b[39m\u001b[33m__main__\u001b[39m\u001b[33m\"\u001b[39m:\n\u001b[32m    255\u001b[39m     \u001b[38;5;66;03m# read_tapestri_H5(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_LPS.dna.h5\")\u001b[39;00m\n\u001b[32m    256\u001b[39m     \u001b[38;5;66;03m# amplicon_matrix = read_dna_count_matrix(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_LPS.dna.h5\")\u001b[39;00m\n\u001b[32m    257\u001b[39m     \u001b[38;5;66;03m# amplicon_level_quality(amplicon_matrix=amplicon_matrix, threshold=5)\u001b[39;00m\n\u001b[32m--> \u001b[39m\u001b[32m259\u001b[39m     df_cell, df_amplicon = \u001b[43mconstruct_amplicon_df\u001b[49m\u001b[43m(\u001b[49m\u001b[43mthreshold\u001b[49m\u001b[43m=\u001b[49m\u001b[32;43m10\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[32m    260\u001b[39m     boxplot_quality(df_cell=df_cell, df_amplicon=df_amplicon)\n\u001b[32m    261\u001b[39m     \u001b[38;5;66;03m# violinplot_quality(df_cell=df_cell, df_amplicon=df_amplicon)\u001b[39;00m\n",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[2]\u001b[39m\u001b[32m, line 90\u001b[39m, in \u001b[36mconstruct_amplicon_df\u001b[39m\u001b[34m(threshold)\u001b[39m\n\u001b[32m     86\u001b[39m pbs_cell_level_qc = (pbs > threshold).sum(axis=\u001b[32m1\u001b[39m)/pbs.shape[\u001b[32m1\u001b[39m]\n\u001b[32m     87\u001b[39m pbs_amplicon_level_qc = (pbs > threshold).sum(axis=\u001b[32m0\u001b[39m)/pbs.shape[\u001b[32m0\u001b[39m]\n\u001b[32m     89\u001b[39m df_cell = pd.DataFrame({\n\u001b[32m---> \u001b[39m\u001b[32m90\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mPercentage\u001b[39m\u001b[33m\"\u001b[39m: np.concat([lps_cell_level_qc, pbs_cell_level_qc, \u001b[43mmixed_cell_level_qc\u001b[49m]),\n\u001b[32m     91\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mCondition\u001b[39m\u001b[33m\"\u001b[39m: [\u001b[33m\"\u001b[39m\u001b[33mLPS\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(lps_cell_level_qc) + [\u001b[33m\"\u001b[39m\u001b[33mPBS\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(pbs_cell_level_qc) + [\u001b[33m\"\u001b[39m\u001b[33mMixed\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(mixed_cell_level_qc)\n\u001b[32m     92\u001b[39m })\n\u001b[32m     94\u001b[39m df_amplicon = pd.DataFrame({\n\u001b[32m     95\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mPercentage\u001b[39m\u001b[33m\"\u001b[39m: np.concat([lps_amplicon_level_qc, pbs_amplicon_level_qc, mixed_amplicon_level_qc]),\n\u001b[32m     96\u001b[39m     \u001b[33m\"\u001b[39m\u001b[33mCondition\u001b[39m\u001b[33m\"\u001b[39m: [\u001b[33m\"\u001b[39m\u001b[33mLPS\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(lps_amplicon_level_qc) + [\u001b[33m\"\u001b[39m\u001b[33mPBS\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(pbs_amplicon_level_qc) + [\u001b[33m\"\u001b[39m\u001b[33mMixed\u001b[39m\u001b[33m\"\u001b[39m]*\u001b[38;5;28mlen\u001b[39m(mixed_amplicon_level_qc)\n\u001b[32m     97\u001b[39m })\n\u001b[32m     98\u001b[39m \u001b[38;5;66;03m# import pdb; pdb.set_trace()\u001b[39;00m\n",
      "\u001b[31mNameError\u001b[39m: name 'mixed_cell_level_qc' is not defined"
     ]
    }
   ],
   "source": [
    "\n",
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "import argparse\n",
    "import seaborn as sns\n",
    "import sys\n",
    "import logging\n",
    "import configparser\n",
    "import anndata\n",
    "import matplotlib.pyplot as plt\n",
    "import re\n",
    "import matplotlib.patches as mpatches\n",
    "import os\n",
    "import argparse\n",
    "import copy\n",
    "import h5py\n",
    "\n",
    "# function reads the loomfile downloaded from Tapestri portal\n",
    "def read_tapestri_H5(filename):\n",
    "    \"\"\"\n",
    "    Read data from MissionBio's formatted loom file.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    filename : str\n",
    "        Path to the loom file (.loom)\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    anndata.AnnData\n",
    "        An anndata object with the following layers:\n",
    "        adata.X: GATK calls\n",
    "        adata.layers['e']: reads with evidence of mutation\n",
    "        adata.layers['no_e']: reads without evidence of mutation\n",
    "    \"\"\"\n",
    "\n",
    "    with h5py.File(filename, \"r\") as file:\n",
    "        # get the variant name, amplicon, chromosome, location]\n",
    "        import pdb; pdb.set_trace()\n",
    "        variant_names = file['assays'][\"dna_variants\"][\"ca\"]['id'][:] # binary\n",
    "        variant_names = np.array([i.decode(\"utf-8\") for i in variant_names])\n",
    "\n",
    "\n",
    "        amplicon_names = file['assays']['dna_variants']['ca']['amplicon'][:] # binary\n",
    "        amplicon_names = np.array([i.decode(\"utf-8\") for i in amplicon_names])\n",
    "\n",
    "        chromosome = file['assays']['dna_variants']['ca']['CHROM'][:] # binary\n",
    "        chromosome = np.array([i.decode(\"utf-8\") for i in chromosome])\n",
    "\n",
    "\n",
    "        location = file['assays']['dna_variants']['ca']['POS'][:]\n",
    "\n",
    "        # get the barcode\n",
    "        barcodes = file['assays']['dna_variants']['ra']['barcode'][:] # binary\n",
    "        barcodes = np.array([i.decode(\"utf-8\") for i in barcodes])\n",
    "\n",
    "        mutation_matrix = file['assays']['dna_variants']['layers']['NGT'][:,:]\n",
    "\n",
    "        adata = anndata.AnnData(X=mutation_matrix, dtype=np.int8)\n",
    "        adata.obs_names = barcodes\n",
    "        adata.var_names = variant_names\n",
    "        adata.varm[\"amplicon\"] = amplicon_names\n",
    "        adata.varm[\"chrom\"] = chromosome\n",
    "        adata.varm[\"loc\"] = location\n",
    "    file.close()\n",
    "\n",
    "    return adata\n",
    "\n",
    "\n",
    "def read_dna_count_matrix(filename):\n",
    "    with h5py.File(filename, \"r\") as file:\n",
    "        amplicon_matrix = file['assays']['dna_read_counts']['layers']['read_counts'][:]\n",
    "    return amplicon_matrix\n",
    "\n",
    "\n",
    "def construct_amplicon_df(threshold=5):\n",
    "\n",
    "    lps = read_dna_count_matrix(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_LPS.dna.h5\")\n",
    "    pbs = read_dna_count_matrix(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_PBS.dna.h5\")\n",
    "    # mixed = read_dna_count_matrix(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_mixed_new.dna.h5\")\n",
    "\n",
    "    lps_cell_level_qc = (lps > threshold).sum(axis=1)/lps.shape[1]\n",
    "    lps_amplicon_level_qc = (lps > threshold).sum(axis=0)/lps.shape[0]\n",
    "\n",
    "    pbs_cell_level_qc = (pbs > threshold).sum(axis=1)/pbs.shape[1]\n",
    "    pbs_amplicon_level_qc = (pbs > threshold).sum(axis=0)/pbs.shape[0]\n",
    "\n",
    "    df_cell = pd.DataFrame({\n",
    "        \"Percentage\": np.concat([lps_cell_level_qc, pbs_cell_level_qc]),\n",
    "        \"Condition\": [\"LPS\"]*len(lps_cell_level_qc) + [\"PBS\"]*len(pbs_cell_level_qc) \n",
    "    })\n",
    "\n",
    "    df_amplicon = pd.DataFrame({\n",
    "        \"Percentage\": np.concat([lps_amplicon_level_qc, pbs_amplicon_level_qc, ]),\n",
    "        \"Condition\": [\"LPS\"]*len(lps_amplicon_level_qc) + [\"PBS\"]*len(pbs_amplicon_level_qc) }\n",
    "    # import pdb; pdb.set_trace()\n",
    "    return df_cell, df_amplicon\n",
    "    \n",
    "\n",
    "def boxplot_quality(df_cell, df_amplicon):\n",
    "    color = \"lightblue\"\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.boxplot(x=\"Condition\", y=\"Percentage\", data=df_cell, color=color)\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    #plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    # plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"cell_level_quality_box.png\", dpi=300)\n",
    "    plt.savefig(\"cell_level_quality_box.svg\", dpi=300)\n",
    "\n",
    "\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.boxplot(x=\"Condition\", y=\"Percentage\", data=df_amplicon, color=color)\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    #plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    # plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    # plt.savefig(\"figures/amplicon_level_quality_box.png\", dpi=300)\n",
    "    # plt.savefig(\"figures/amplicon_level_quality_box.svg\", dpi=300)\n",
    "    plt.show()\n",
    "\n",
    "\n",
    "def violinplot_quality(df_cell, df_amplicon):\n",
    "    color = \"lightblue\"\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.violinplot(x=\"Condition\", y=\"Percentage\", data=df_cell, color=color)\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    #plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    # plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"cell_level_quality_violin.png\", dpi=300)\n",
    "    plt.savefig(\"cell_level_quality_violin.svg\", dpi=300)\n",
    "\n",
    "\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.violinplot(x=\"Condition\", y=\"Percentage\", data=df_amplicon, color=color)\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    #plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    # plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"figures/amplicon_level_quality_violin.png\", dpi=300)\n",
    "    plt.savefig(\"figures/amplicon_level_quality_violin.svg\", dpi=300)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "def cell_level_quality(amplicon_matrix, threshold=5,):\n",
    "    # count for each cell, the percentage of amplicon that pass the thershold.\n",
    "    cell_level_passed_amplicon = (amplicon_matrix > threshold).sum(axis=1)/amplicon_matrix.shape[1]\n",
    "    df = pd.DataFrame({\n",
    "    \"Percentage\": cell_level_passed_amplicon,\n",
    "    \"Condition\": [\"121 complicon\"] * len(cell_level_passed_amplicon)\n",
    "    })\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.violinplot(x=\"Condition\", y=\"Percentage\", data=df, color=\"skyblue\")\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    plt.ylim(0.6, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    plt.title(\"violin for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"cell_level_quality_violin.png\", dpi=300)\n",
    "    plt.savefig(\"cell_level_quality_violin.svg\", dpi=300)\n",
    "\n",
    "\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.boxplot(x=\"Condition\", y=\"Percentage\", data=df, color=\"skyblue\")\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    plt.ylim(0.6, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"cell_level_quality_box.png\", dpi=300)\n",
    "    plt.savefig(\"cell_level_quality_box.svg\", dpi=300)\n",
    "\n",
    "        \n",
    "def amplicon_level_quality(amplicon_matrix, threshold=5):\n",
    "    # count for each cell, the percentage of amplicon that pass the thershold.\n",
    "    amplicon_level_passed_amplicon = (amplicon_matrix > threshold).sum(axis=0)/amplicon_matrix.shape[0]\n",
    "    # import pdb; pdb.set_trace()\n",
    "    df = pd.DataFrame({\n",
    "    \"Percentage\": amplicon_level_passed_amplicon,\n",
    "    \"Condition\": [\"121 complicon\"] * len(amplicon_level_passed_amplicon)\n",
    "    })\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.violinplot(x=\"Condition\", y=\"Percentage\", data=df, color=\"skyblue\")\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    plt.title(\"violin for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"amplicon_level_quality_violin.png\", dpi=300)\n",
    "    plt.savefig(\"amplicon_level_quality_violin.svg\", dpi=300)\n",
    "\n",
    "\n",
    "    # Create a boxplot\n",
    "    plt.figure(figsize=(8, 6))\n",
    "    sns.boxplot(x=\"Condition\", y=\"Percentage\", data=df, color=\"skyblue\")\n",
    "\n",
    "    # Ensure the y-axis is set from 0 to 1\n",
    "    plt.ylim(0, 1)\n",
    "\n",
    "    # Add labels and title\n",
    "    plt.xlabel(\"Condition\")\n",
    "    plt.ylabel(\"Percentage\")\n",
    "    plt.title(\"Boxplot for 121 complicon condition\")\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(\"amplicon_level_quality_box.png\", dpi=300)\n",
    "    plt.savefig(\"amplicon_level_quality_box.svg\", dpi=300)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    # read_tapestri_H5(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_LPS.dna.h5\")\n",
    "    # amplicon_matrix = read_dna_count_matrix(\"/mnt/data/project/stag_analysis/data/wholeTranscriptomics/WholeTranscriptomics_LPS.dna.h5\")\n",
    "    # amplicon_level_quality(amplicon_matrix=amplicon_matrix, threshold=5)\n",
    "\n",
    "    df_cell, df_amplicon = construct_amplicon_df(threshold=10)\n",
    "    boxplot_quality(df_cell=df_cell, df_amplicon=df_amplicon)\n",
    "    # violinplot_quality(df_cell=df_cell, df_amplicon=df_amplicon)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pid",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
