{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"machine_shape":"hm","authorship_tag":"ABX9TyM3kzAq+xyhSUBQmN5T+dRp"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"0LgAKlLiO2ge"},"outputs":[],"source":["%pip install scanpy\n","%pip install anndata\n","%pip install MuData\n","%pip install loompy"]},{"cell_type":"code","source":["from mudata import MuData\n","import numpy as np\n","from tqdm import tqdm\n","import pandas as pd\n","import anndata\n","import scanpy\n","import csv\n","import sys\n","import loompy\n","import logging\n","import os\n","from matplotlib import pyplot as plt\n","import plotly.graph_objects as go"],"metadata":{"id":"9DuFYF2CO8V0"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# Configure logging\n","logging.basicConfig(stream=sys.stdout, level=logging.INFO, format='%(message)s')\n","logger = logging.getLogger(__name__)\n","# dont judge me too much...\n","import warnings\n","warnings.simplefilter(action='ignore', category=Warning)"],"metadata":{"id":"liThvx3XPBbV"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from ast import Continue\n","# function reads the loomfile downloaded from Tapestri portal\n","def read_tapestri_loom(filename):\n","    \"\"\"\n","    Read data from MissionBio's formatted loom file.\n","\n","    Parameters\n","    ----------\n","    filename : str\n","        Path to the loom file (.loom)\n","\n","    Returns\n","    -------\n","    anndata.AnnData\n","        An anndata object with the following layers:\n","        adata.X: GATK calls\n","        adata.layers['e']: reads with evidence of mutation\n","        adata.layers['no_e']: reads without evidence of mutation\n","    \"\"\"\n","    loom_file = loompy.connect(filename)\n","\n","    variant_names, amplicon_names, chromosome, location = (\n","        loom_file.ra['id'], loom_file.ra['amplicon'], loom_file.ra['CHROM'], loom_file.ra['POS'])\n","\n","    barcodes = [barcode.split('-')[0] for barcode in loom_file.ca['barcode']]\n","    adata = anndata.AnnData(np.transpose(loom_file[:, :]), dtype=loom_file[:, :].dtype)\n","    adata.layers['e'] = np.transpose(loom_file.layers['AD'][:, :])\n","    adata.layers['no_e'] = np.transpose(loom_file.layers['RO'][:, :])\n","    adata.var_names = variant_names\n","    adata.obs_names = barcodes\n","    adata.varm['amplicon'] = amplicon_names\n","    adata.varm['chrom'] = chromosome\n","    adata.varm['loc'] = location\n","\n","    loom_file.close()\n","\n","    return adata\n","\n","def read_HyPR_file(HyPR_file):\n","\n","    HyPR_array = np.loadtxt(HyPR_file, dtype=str, skiprows=1)\n","    barcodes = HyPR_array[:, 0]\n","    with open(HyPR_file, 'r') as f:\n","        first_line = f.readline().strip()\n","        values = first_line.split('\\t')\n","        var_names = np.array(values, dtype=str)[:]\n","    X = np.asarray(HyPR_array[:, 1:], dtype=int)\n","    if np.size(var_names) != X.shape[1]: # sloppy I shouldn't need this\n","      var_names = var_names[1:]\n","    df = pd.DataFrame(X, columns=var_names)\n","    df.columns = df.columns.str.split('_').str[0]\n","    collapsed_df = df.groupby(axis=1, level=0).sum()\n","    collapsed_X = collapsed_df.to_numpy()\n","    collapsed_var_names = collapsed_df.columns.to_numpy()\n","\n","    adata_hypr = anndata.AnnData(collapsed_X)\n","    adata_hypr.obs_names = barcodes\n","\n","    try:\n","      adata_hypr.var_names = collapsed_var_names\n","    except:\n","      adata_hypr.var_names = collapsed_var_names[1:]\n","\n","    return adata_hypr\n","\n","# Function for finding the intersecting barcodes between the modalities\n","def find_intersecting(mdata):\n","    \"\"\"\n","    Find intersecting barcodes and add 'intersecting' to obs with boolean values.\n","\n","    Parameters\n","    ----------\n","    mdata : object\n","        The mudata object.\n","\n","    Raises\n","    ------\n","    AssertionError\n","        If the length of mod_names is not equal to 2.\n","    \"\"\"\n","\n","    mod_names = list(mdata.mod.keys())\n","    assert len(mod_names) == 2, 'Function not implemented for mod_names with length different from 2'\n","\n","    obs_1, obs_2 = mdata[mod_names[0]].obs_names, mdata[mod_names[1]].obs_names\n","\n","    cmn_barcodes, idx_1, idx_2 = np.intersect1d(obs_1, obs_2, return_indices=True)\n","\n","    # Initializing 'intersecting' columns to False\n","    mdata[mod_names[0]].obs['intersecting'] = False\n","    mdata[mod_names[1]].obs['intersecting'] = False\n","\n","    # Updating 'intersecting' columns where True\n","    mdata[mod_names[0]].obs['intersecting'].iloc[idx_1] = True\n","    mdata[mod_names[1]].obs['intersecting'].iloc[idx_2] = True\n","\n","    logger.info(f\"Found {idx_1.size} intersecting barcodes\")\n","\n","\n","def annotate_genotype(mdata, variants, window=4, obs_key='mutant_type',\n","                      genotype_key='tapestri', phenotype_key='HyPR',\n","                      ignore_bystanders=False, write_all_as=False,\n","                      ignore_mixed=False, only_type=None, germline_variants=None):\n","\n","    adata = mdata.mod[genotype_key]\n","\n","    # loop through the variants in the variant list\n","    for variant in variants:\n","\n","        if obs_key in mdata.mod[phenotype_key].obs:\n","          pass\n","        else:\n","          mdata.mod[phenotype_key].obs[obs_key] = 'unannotated'\n","          mdata.mod[genotype_key].obs[obs_key] = 'unannotated'\n","\n","        try:\n","            idx = adata.var_names.get_loc(variant)\n","        except KeyError:\n","            logger.info(str(variant) + ' was not found in the genotype modality anndata')\n","            continue\n","\n","        # identify which cells have the variant\n","        bkg_idx, het_idx, hom_idx = [np.flatnonzero(adata.X[:, idx] == i) for i in (0, 1, 2)]\n","\n","        if only_type == 'het':\n","          hom_idx = None\n","        if only_type == 'hom':\n","          het_idx = None\n","\n","        if ignore_bystanders == True:\n","\n","          if het_idx is not None:\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[het_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            if write_all_as == False:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_het'\n","              mdata.mod[genotype_key].obs[obs_key][het_idx]= str(variant) + '_het'\n","            else:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = write_all_as\n","\n","          if hom_idx is not None:\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[hom_idx],\n","                                                                   mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            if write_all_as == False:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype]= str(variant) + '_hom'\n","              mdata.mod[genotype_key].obs[obs_key][hom_idx]= str(variant) + '_hom'\n","            else:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = write_all_as\n","\n","        else:\n","\n","          # identify indices of nearby SNPs that are within the bystander window\n","          chromosome, location = adata.varm['chrom'][idx], adata.varm['loc'][idx]\n","          nearby_ = np.where(adata.varm['chrom'] == chromosome)\n","          idx_ = np.where(np.logical_and(np.asarray(adata.varm['loc'], dtype=int) < int(location + window), np.asarray(adata.varm['loc'], dtype=int) > int(location - window)))\n","          nearby = np.intersect1d(nearby_, idx_)\n","\n","          # drop wildcards\n","          np.delete(nearby, [i for i, var_name in enumerate(adata.var_names[nearby]) if var_name.endswith('*')])\n","\n","          # drop germline variants from bystander analysis\n","          if germline_variants is not None:\n","            print(\"dropped germline\")\n","            mask = [name in germline_variants for name in adata.var_names]\n","            germ_indices = np.where(mask)[0]\n","            bad_idxs = np.intersect1d(germ_indices, nearby)\n","            nearby = np.setdiff1d(nearby, bad_idxs)\n","\n","          mut_vals = adata.X[:, nearby]\n","          mut_vals[mut_vals == 3] = 0\n","          mut_vals[mut_vals > 1] = 1\n","          sum_muts = np.sum(mut_vals, axis=1)\n","\n","          if het_idx is not None:\n","            pure_het_idx = np.intersect1d(np.argwhere(sum_muts == 1), het_idx)\n","            not_pure_het_idx = np.intersect1d(np.argwhere(sum_muts != 1), het_idx)\n","            mdata.mod[genotype_key].obs[obs_key][pure_het_idx] = str(variant) + '_het_pure'\n","            mdata.mod[genotype_key].obs[obs_key][not_pure_het_idx] = str(variant) + '_het_bystander'\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[not_pure_het_idx],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype]= str(variant) + '_het_bystander'\n","            logger.info(f\"Annotated {idx_genotype.size} heterozygous mutants with bystanders\")\n","\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[pure_het_idx],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_het_pure'\n","            logger.info(f\"Annotated {idx_genotype.size} pure heterozygous mutants\")\n","\n","\n","          if hom_idx is not None:\n","\n","            pure_hom_idx = np.intersect1d(np.argwhere(sum_muts == 1), hom_idx)\n","            not_pure_hom_idx = np.intersect1d(np.argwhere(sum_muts != 1), hom_idx)\n","\n","            mdata.mod[genotype_key].obs[obs_key][pure_hom_idx] = str(variant) + '_hom_pure'\n","            mdata.mod[genotype_key].obs[obs_key][not_pure_hom_idx] = str(variant) + '_hom_bystander'\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[not_pure_hom_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_hom_bystander'\n","            logger.info(f\"Annotated {idx_genotype.size} homozygous mutants with bystanders\")\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[pure_hom_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_hom_pure'\n","            logger.info(f\"Annotated {idx_genotype.size} pure homozygous mutants\")\n","\n","\n","    if ignore_mixed == False:\n","      # first, lets find the cells that have mixed edits of the intended variants\n","      indices = []\n","      for variant in variants:\n","        if variant in germline_variants:\n","          continue\n","\n","        try:\n","            indices.append(adata.var_names.get_loc(variant))\n","        except KeyError:\n","            logger.info(str(variant) + ' was not found in the genotype modality anndata')\n","            continue\n","      idxs = np.asarray(indices).flatten()\n","\n","      print(idxs)\n","\n","      edit_array = np.zeros(adata.X[:, idxs].shape)\n","      edit_array[adata.X[:, idxs] == 1] = 1\n","      edit_array[adata.X[:, idxs] == 2] = 1\n","      mixed_idxs = np.argwhere(np.sum(edit_array, axis=1) >= 2).ravel()\n","\n","      mdata.mod[genotype_key].obs[obs_key][mixed_idxs] = 'mixed_mutant'\n","      cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[mixed_idxs],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","      mdata.mod[phenotype_key].obs[obs_key][idx_phenotype]= 'mixed_mutant'\n","\n","def call_germline(tapestri_adata, cutoff=0.25):\n","\n","  fraction_het = np.mean(tapestri_adata.X == 1, axis=0)\n","  het_germline = tapestri_adata.var_names[fraction_het > cutoff]\n","\n","  fraction_hom = np.mean(tapestri_adata.X == 2, axis=0)\n","  hom_germline = tapestri_adata.var_names[fraction_hom > cutoff]\n","\n","  return list(het_germline), list(hom_germline)"],"metadata":{"id":"B_i96WlXPCYd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["het_germline, hom_germline = call_germline(mdata_list[1].mod['tapestri'], cutoff=0.3)\n","germline = het_germline + hom_germline"],"metadata":{"id":"eW0NmojL6AQj"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import seaborn as sns\n","def barcode_rank_plot(adata, minimum=0, xmax=None):\n","\n","  # Sum the UMIs for each cell\n","  cell_umi_counts_all = adata.X.sum(axis=1)\n","  cell_idxs = np.argwhere(cell_umi_counts_all > minimum)\n","  cell_umi_counts = cell_umi_counts_all[cell_idxs]\n","\n","  # Convert to numpy array if it's not already\n","  cell_umi_counts = np.array(cell_umi_counts).flatten()\n","\n","  # Sort the UMI counts in descending order for the rank plot\n","  sorted_umi_counts = np.sort(cell_umi_counts)[::-1]\n","\n","  plt.figure(figsize=(5, 4), dpi=150)\n","  sns.lineplot(x=range(1, len(sorted_umi_counts) + 1), y=sorted_umi_counts)\n","  plt.xlabel('Barcode Rank')\n","  plt.ylabel('UMI Count')\n","  plt.title('Cell Barcode Rank Plot')\n","  plt.yscale('log')  # Log scale for better visualization\n","  if xmax is not None:\n","    plt.xlim(0, xmax)\n","  plt.show()"],"metadata":{"id":"NQDVPavoPGHU"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from google.colab import drive\n","drive.mount('/content/drive')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"v9eqMoF_PI3-","executionInfo":{"status":"ok","timestamp":1714617117692,"user_tz":240,"elapsed":20025,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"}},"outputId":"e3da6b9e-b5d3-4f75-affa-c1d9e008bd95"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}]},{"cell_type":"code","source":["# define where all the files are on the drive\n","data_dir = 'path/to/dsifng/data'\n","\n","file_locations = [['tapestri_loom_files/MDM_DSIFNG_Ctrl_gDNA1.cells.loom', 'HyPR_probe_matrices/MDM_DSIFNG_Ctrl_HyPR1.txt'],\n","                  ['tapestri_loom_files/MDM_DSIFNG_Ctrl_gDNA2.cells.loom', 'HyPR_probe_matrices/MDM_DSIFNG_Ctrl_HyPR2.txt'],\n","                  ['tapestri_loom_files/MDM_DSIFNG_IFNg_gDNA1.cells.loom', 'HyPR_probe_matrices/MDM_DSIFNG_IFNg_HyPR1.txt'],\n","                  ['tapestri_loom_files/MDM_DSIFNG_IFNg_gDNA2.cells.loom', 'HyPR_probe_matrices/MDM_DSIFNG_IFNg_HyPR2.txt'],\n","                  ['tapestri_loom_files/MDM_DSIFNG_IFNg_gDNA3.cells.loom', 'HyPR_probe_matrices/MDM_DSIFNG_IFNg_HyPR3.txt']]\n","\n","sample_names = ['Ctrl1', 'Ctrl2', 'IFNg1', 'IFNg2', 'IFNg3']\n","\n","# check to make sure all those files exist\n","for i in tqdm(range(0, len(file_locations))):\n","  tap_file = data_dir + file_locations[i][0]\n","  hypr_file = data_dir + file_locations[i][1]\n","  if os.path.exists(tap_file):\n","      pass\n","  else:\n","    print(tap_file)\n","  if os.path.exists(hypr_file):\n","    pass\n","  else:\n","    print(hypr_file)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"mcG2R-otPJ7H","executionInfo":{"status":"ok","timestamp":1714617127177,"user_tz":240,"elapsed":875,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"}},"outputId":"c136c0af-7243-4bbe-c836-344eafdc784b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 5/5 [00:00<00:00,  6.59it/s]\n"]}]},{"cell_type":"code","source":["mdata_list = []\n","for i in tqdm(range(0, len(file_locations))):\n","    adata_hypr = read_HyPR_file(data_dir + file_locations[i][1])\n","    adata_tapestri = read_tapestri_loom(data_dir + file_locations[i][0])\n","    mdata_list.append(MuData({'HyPR': adata_hypr, 'tapestri': adata_tapestri}))"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"_7SdHiXEPxpF","executionInfo":{"status":"ok","timestamp":1714617438470,"user_tz":240,"elapsed":308261,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"}},"outputId":"64a4ed20-3e69-468e-cb2e-216c050e242b"},"execution_count":null,"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 5/5 [05:08<00:00, 61.70s/it]\n"]}]},{"cell_type":"code","source":["sns.set_theme()\n","for i, sample_name in enumerate(sample_names):\n","  print(sample_name)\n","  barcode_rank_plot(mdata_list[i].mod['HyPR'])"],"metadata":{"id":"uyy3s7A0P_0W"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["for i, sample_name in enumerate(sample_names):\n","  scanpy.pp.filter_cells(mdata_list[i].mod['HyPR'], min_counts=1000)"],"metadata":{"id":"MqmU9pKtQCoA"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["for i in range(0, len(mdata_list)):\n","  adata = mdata_list[i].mod['tapestri']\n","  df = pd.DataFrame(adata.X, index=adata.obs_names, columns=adata.var_names)\n","  df.to_csv('/content/drive/MyDrive/STAG_Seq_Folder/DSIFNG/GATK_' + sample_names[i] + '.txt', sep='\\t', index=True, header=True)"],"metadata":{"id":"JQT_5WLku5FG"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# find intersecting and remove any cells that are not intersecting across the modalities since we wont need them\n","for mdata in mdata_list:\n","  find_intersecting(mdata)\n","  mdata.mod['tapestri'] = mdata.mod['tapestri'][mdata.mod['tapestri'].obs['intersecting'] == True].copy()\n","  mdata.mod['HyPR'] = mdata.mod['HyPR'][mdata.mod['HyPR'].obs['intersecting'] == True].copy()\n","\n","  # re-order the observations\n","  hypr_order = mdata.mod['HyPR'].obs_names\n","  mdata.mod['tapestri'] = mdata.mod['tapestri'][hypr_order]"],"metadata":{"id":"asYkQDFvRo3W"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# identify the AAVS\n","chrom, loc, edit_type = [], [], []\n","variant_names = np.asarray(mdata.mod['tapestri'].var_names.values)\n","for i in range(0, np.size(variant_names)):\n","    chrom.append(variant_names[i].split(':')[0])\n","    loc.append(int(variant_names[i].split(':')[1]))\n","    edit_type.append(variant_names[i].split(':')[2])\n","\n","control_editing = []\n","for i in range(0, len(chrom)):\n","    if chrom[i] == 'chr19':\n","        if loc[i] in range(55115745, 55115764):\n","            if edit_type[i] == \"A/G\":\n","                control_editing.append(variant_names[i])\n","            if edit_type[i] == \"T/C\":\n","                control_editing.append(variant_names[i])\n","        if loc[i] in range(55115752, 55115771):\n","            if edit_type[i] == \"C/T\":\n","                control_editing.append(variant_names[i])\n","            if edit_type[i] == \"G/A\":\n","                control_editing.append(variant_names[i])"],"metadata":{"id":"FxF-ah3vRqjs"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# identify the AAVS\n","chrom, loc, edit_type = [], [], []\n","variant_names = np.asarray(mdata.mod['tapestri'].var_names.values)\n","for i in range(0, np.size(variant_names)):\n","    chrom.append(variant_names[i].split(':')[0])\n","    loc.append(int(variant_names[i].split(':')[1]))\n","    edit_type.append(variant_names[i].split(':')[2])"],"metadata":{"id":"MrimixEPTqV1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["variants = list(np.loadtxt(data_dir + 'var_list.txt', dtype=str))"],"metadata":{"id":"Keho8Xlpki9b"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["variants.remove('chr21:33415005:A/G')"],"metadata":{"id":"YUfVKgAm6qK9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["ignore_bystander_list = list(np.loadtxt(data_dir + 'ignore_bystanders_vars.txt', dtype=str))"],"metadata":{"id":"uJMA4OgnDV4_"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["ptcs = [\"chr6:137207079:A/G\", \"chr21:33421686:G/A\", \"chr1:64873525:A/G\"]"],"metadata":{"id":"x8XD7z-3SbWC"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["for i, mdata in enumerate(mdata_list):\n","    annotate_genotype(mdata, control_editing, ignore_bystanders=True, ignore_mixed=True, write_all_as='AAVS')\n","    annotate_genotype(mdata, variants, ignore_bystanders=False, ignore_mixed=False, window=10, germline_variants=['chr21:33415005:A/G'])\n","    annotate_genotype(mdata, ignore_bystander_list, ignore_bystanders=False, ignore_mixed=True, window=10, germline_variants=['chr21:33415005:A/G'])\n","    annotate_genotype(mdata, ptcs, ignore_bystanders=True, ignore_mixed=True)"],"metadata":{"id":"hWaYclulRrS1"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["for i, mdata in enumerate(mdata_list):\n","  print(np.argwhere(mdata.mod['HyPR'].obs['mutant_type'] == 'AAVS').shape)"],"metadata":{"id":"0mqAN5yrGSUl"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["mdata_list"],"metadata":{"id":"n0nJbTxpydcq"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["from sklearn.cluster import KMeans\n","def cluster_genotypes(mdata_list, n_genotypes, frequency_cutoff=0.1):\n","\n","  adatas_listed = [mdata.mod['tapestri'] for mdata in mdata_list]\n","  temp_adata = anndata.concat(adatas_listed, join='inner', label=\"batch\")\n","\n","  fraction_het = np.mean(temp_adata.X == 1, axis=0)\n","  het_germline = list(temp_adata.var_names[fraction_het > frequency_cutoff])\n","\n","  fraction_hom = np.mean(temp_adata.X == 2, axis=0)\n","  hom_germline = list(temp_adata.var_names[fraction_hom > frequency_cutoff])\n","\n","  temp_adata = temp_adata[:, het_germline + hom_germline]\n","\n","  temp_adata.layers['norm_e'] = temp_adata.layers['e'] / np.sum((temp_adata.layers['e'] + temp_adata.layers['no_e'] + 1), axis=1)[:, np.newaxis]\n","\n","  scanpy.pp.pca(temp_adata, layer='norm_e')\n","  kmeans = KMeans(n_clusters=n_genotypes, random_state=0).fit(temp_adata.layers['norm_e'])\n","\n","  cluster_labels = kmeans.labels_\n","  temp_adata.obs['cluster_labels'] = cluster_labels\n","\n","  # scanpy.pl.pca(temp_adata, color=cluster_labels)\n","\n","  return cluster_labels, temp_adata"],"metadata":{"id":"7ppetM3DVJj9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["labels, temp_adata = cluster_genotypes(mdata_list, 2)"],"metadata":{"id":"huPckmrWVNQd"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["adatas_listed = [mdata.mod['tapestri'] for mdata in mdata_list]\n","cat = anndata.concat(adatas_listed, join='inner', label=\"batch\")\n","cat.layers['vaf'] = cat.layers['e'] #/ (cat.layers['e'] + cat.layers['no_e'] + 1)"],"metadata":{"id":"d1oNCEB4pXw6"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["scanpy.pp.normalize_total(cat, layer='vaf')"],"metadata":{"id":"cmdutgIVrw4L"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# create a concatenated anndata object for the HyPR\n","adata_hypr_list = [mdata.mod['HyPR'] for mdata in mdata_list]\n","\n","for anndata_obj, batch_name in zip(adata_hypr_list, sample_names):\n","    anndata_obj.obs['batch'] = batch_name\n","\n","concatenated_anndata = anndata.concat(adata_hypr_list)\n","\n","concatenated_anndata.obs['genotype_cluster'] = labels\n","\n","concatenated_anndata.obs['stimulation'] = concatenated_anndata.obs['batch'].str.slice(0, 4).copy()"],"metadata":{"id":"FYAfNGmaVOvO"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["adata_tap_list = [mdata.mod['tapestri'] for mdata in mdata_list]\n","\n","concatenated_anndata = anndata.concat(adata_tap_list)"],"metadata":{"id":"gRNiOBHESDar"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["concatenated_anndata.obs['genotype_cluster'] = labels"],"metadata":{"id":"ylWPYzbqR9d9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["concatenated_anndata.layers['raw'] = concatenated_anndata.X.copy()"],"metadata":{"id":"MJhTxnzaVvIv"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["scanpy.pp.normalize_total(concatenated_anndata, target_sum=1e4)\n","scanpy.pp.log1p(concatenated_anndata)"],"metadata":{"id":"UOpkO2vnV4w9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["scanpy.tl.pca(concatenated_anndata)\n","scanpy.pp.neighbors(concatenated_anndata)\n","scanpy.tl.umap(concatenated_anndata, min_dist=0.5, spread=3)"],"metadata":{"id":"68ExQDP8V696"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["donor0_subset = concatenated_anndata[concatenated_anndata.obs['genotype_cluster'] == 0]\n","donor1_subset = concatenated_anndata[concatenated_anndata.obs['genotype_cluster'] == 1]"],"metadata":{"id":"w55RuI_EumII"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# lets save this as an .h5ad for easy use in Python\n","output_path = \"where/to/write/h5ad\"\n","concatenated_anndata.write(output_path)"],"metadata":{"id":"roCSWwq4V8rL"},"execution_count":null,"outputs":[]}]}