{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"0pm7O-8Txii_"},"outputs":[],"source":["%pip install scanpy\n","%pip install anndata\n","%pip install MuData\n","%pip install loompy"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":15695,"status":"ok","timestamp":1718727767392,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"rErajClHxnMF","outputId":"9bb38d63-820f-4b10-99af-24be80b5945f"},"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:68: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def twobit_to_dna(twobit: int, size: int) -> str:\n","/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:85: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def dna_to_twobit(dna: str) -> int:\n","/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:102: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def twobit_1hamming(twobit: int, size: int) -> List[int]:\n"]}],"source":["from mudata import MuData\n","import numpy as np\n","from tqdm import tqdm\n","import pandas as pd\n","import anndata\n","import scanpy\n","import csv\n","import sys\n","import loompy\n","import logging\n","import os\n","from matplotlib import pyplot as plt"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":19074,"status":"ok","timestamp":1718727786463,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"3I7ToAHBxw3W","outputId":"9e1a7952-3a88-4577-83fd-33ce56028971"},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["# mount drive if neccessary\n","# from google.colab import drive\n","# drive.mount('/content/drive')"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"0WhqGiVPxxE5"},"outputs":[],"source":["h5ad_path = \"path/to/my/processed/h5ad\"\n","adata = scanpy.read_h5ad(h5ad_path)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":5,"status":"ok","timestamp":1718727798734,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"ecauRLBmyKEI","outputId":"beffdfc5-69fd-4926-f296-fe0484f2e23a"},"outputs":[{"output_type":"execute_result","data":{"text/plain":["AnnData object with n_obs × n_vars = 56783 × 1276\n","    obs: 'n_counts', 'intersecting', 'mutant_type', 'batch', 'genotype_cluster', 'stimulation'\n","    uns: 'batch_colors', 'log1p', 'neighbors', 'pca', 'umap'\n","    obsm: 'X_pca', 'X_umap'\n","    varm: 'PCs'\n","    layers: 'raw'\n","    obsp: 'connectivities', 'distances'"]},"metadata":{},"execution_count":6}],"source":["# extract a specific donor\n","adata = adata[adata.obs['genotype_cluster'] == 0] # or 1"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"f73gEe6p5uwE"},"outputs":[],"source":["output_path = \"where/to/write/mutant_de_file.txt\"\n","\n","stims = np.unique(adata.obs['stimulation'].values)\n","all_variants = np.unique(adata.obs['mutant_type'].values)\n","\n","results = {}\n","for i, stim in enumerate(stims):\n","\n","    tmp_adata = adata[adata.obs['stimulation'] == stim].copy()\n","    valid_groups = [group for group in all_variants if np.sum(tmp_adata.obs['mutant_type'] == group) > 1]\n","    scanpy.tl.rank_genes_groups(tmp_adata, 'mutant_type', groups=valid_groups, reference='AAVS', method='wilcoxon')\n","\n","    for j, variant in enumerate(all_variants):\n","        key = f\"{stim}_{variant}\"\n","        if variant not in tmp_adata.uns['rank_genes_groups']['pvals_adj'].dtype.names:\n","            continue\n","        names = tmp_adata.uns['rank_genes_groups']['names'][variant]\n","        pvals_adj = tmp_adata.uns['rank_genes_groups']['pvals_adj'][variant]\n","        logfoldchanges = tmp_adata.uns['rank_genes_groups']['logfoldchanges'][variant]\n","\n","        for gene, pval, logfc in zip(names, pvals_adj, logfoldchanges):\n","            if gene not in results:\n","                results[gene] = {}\n","            results[gene][f\"{key}_pval\"] = pval\n","            results[gene][f\"{key}_logfc\"] = logfc"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"Mz-erK6S54S4"},"outputs":[],"source":["output_path = \"where/to/write/stim_de_file.txt\"\n","\n","# first we pull all the 'unannotated' from mutant_type since we shouldn't be comparing edited cells for this\n","subdata = adata[adata.obs['mutant_type'] == 'unannotated']\n","\n","# use scanpy to do the DE\n","scanpy.tl.rank_genes_groups(subdata, 'stimulation', reference='Ctrl', method='wilcoxon')\n","\n","# simulations with the exception of control stimulation\n","all_stims = [stim for stim in subdata.obs['stimulation'].cat.categories if stim != 'Ctrl']\n","\n","n_genes = len(subdata.var_names)\n","n_stims = len(all_stims)\n","\n","# Initialize matrices for adjusted p-values and logFC\n","pval_matrix = np.full((n_genes, n_stims), np.nan)  # Use np.nan to indicate missing values\n","logfc_matrix = np.full((n_genes, n_stims), np.nan)\n","\n","# Create a dictionary to map gene names to their indices in subdata.var_names\n","gene_to_index = {gene: index for index, gene in enumerate(subdata.var_names)}\n","\n","# loop through and population the p_val matrix and logfc matrix from the DE results\n","for j, stim in enumerate(all_stims):\n","    de_genes = subdata.uns['rank_genes_groups']['names'][stim]\n","    pvals_adj = subdata.uns['rank_genes_groups']['pvals_adj'][stim]\n","    logfc = subdata.uns['rank_genes_groups']['logfoldchanges'][stim]\n","\n","    for de_gene, pval, lf in zip(de_genes, pvals_adj, logfc):\n","        idx = gene_to_index[de_gene]\n","        pval_matrix[idx, j] = pval\n","        logfc_matrix[idx, j] = lf\n","\n","# Write matrices to .txt file\n","with open(output_path, \"w\") as f:\n","    # Header\n","    headers = [\"gene_names\"] + [f\"{stim}_{metric}\" for stim in all_stims for metric in ['pval_adj', 'logfc']]\n","    f.write(\"\\t\".join(headers) + \"\\n\")\n","\n","    for i, gene in enumerate(subdata.var_names):\n","        values = []\n","        for j in range(n_stims):\n","            values.extend([str(pval_matrix[i, j]), str(logfc_matrix[i, j])])\n","        f.write(gene + \"\\t\" + \"\\t\".join(values) + \"\\n\")"]}],"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyOiQax0NWhMW8WOgDPTUrLY"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}