{"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"k-V15-rT8hdP"},"outputs":[],"source":["%pip install scanpy\n","%pip install anndata\n","%pip install MuData\n","%pip install loompy"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":11342,"status":"ok","timestamp":1712154648232,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"fvBREJd28pnQ","outputId":"6b147b8d-aa4a-4457-e3ff-b25989522826"},"outputs":[{"output_type":"stream","name":"stderr","text":["/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:68: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def twobit_to_dna(twobit: int, size: int) -> str:\n","/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:85: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def dna_to_twobit(dna: str) -> int:\n","/usr/local/lib/python3.10/dist-packages/loompy/bus_file.py:102: NumbaDeprecationWarning: The 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\n","  def twobit_1hamming(twobit: int, size: int) -> List[int]:\n"]}],"source":["from mudata import MuData\n","import numpy as np\n","from tqdm import tqdm\n","import pandas as pd\n","import anndata\n","import scanpy\n","import csv\n","import sys\n","import loompy\n","import logging\n","import os\n","from matplotlib import pyplot as plt\n","import plotly.graph_objects as go"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"_ZDXC3Nx87l3"},"outputs":[],"source":["# Configure logging\n","logging.basicConfig(stream=sys.stdout, level=logging.INFO, format='%(message)s')\n","logger = logging.getLogger(__name__)\n","# dont judge me too much...\n","import warnings\n","warnings.simplefilter(action='ignore', category=Warning)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"a46o9grZ9B-g"},"outputs":[],"source":["# function reads the loomfile downloaded from Tapestri portal\n","def read_tapestri_loom(filename):\n","    \"\"\"\n","    Read data from MissionBio's formatted loom file.\n","\n","    Parameters\n","    ----------\n","    filename : str\n","        Path to the loom file (.loom)\n","\n","    Returns\n","    -------\n","    anndata.AnnData\n","        An anndata object with the following layers:\n","        adata.X: GATK calls\n","        adata.layers['e']: reads with evidence of mutation\n","        adata.layers['no_e']: reads without evidence of mutation\n","    \"\"\"\n","    loom_file = loompy.connect(filename)\n","\n","    variant_names, amplicon_names, chromosome, location = (\n","        loom_file.ra['id'], loom_file.ra['amplicon'], loom_file.ra['CHROM'], loom_file.ra['POS'])\n","\n","    barcodes = [barcode.split('-')[0] for barcode in loom_file.ca['barcode']]\n","    adata = anndata.AnnData(np.transpose(loom_file[:, :]), dtype=loom_file[:, :].dtype)\n","    adata.layers['e'] = np.transpose(loom_file.layers['AD'][:, :])\n","    adata.layers['no_e'] = np.transpose(loom_file.layers['RO'][:, :])\n","    adata.var_names = variant_names\n","    adata.obs_names = barcodes\n","    adata.varm['amplicon'] = amplicon_names\n","    adata.varm['chrom'] = chromosome\n","    adata.varm['loc'] = location\n","\n","    loom_file.close()\n","\n","    return adata\n","\n","def read_HyPR_file(HyPR_file, byprobe=False):\n","\n","    HyPR_array = np.loadtxt(HyPR_file, dtype=str, skiprows=1)\n","    barcodes = HyPR_array[:, 0]\n","    with open(HyPR_file, 'r') as f:\n","        first_line = f.readline().strip()\n","        values = first_line.split('\\t')\n","        var_names = np.array(values, dtype=str)[:]\n","    X = np.asarray(HyPR_array[:, 1:], dtype=int)\n","    if np.size(var_names) != X.shape[1]: # sloppy I shouldn't need this\n","      var_names = var_names[1:]\n","    df = pd.DataFrame(X, columns=var_names)\n","\n","    if byprobe == True:\n","      adata_hypr = anndata.AnnData(X)\n","      adata_hypr.obs_names = barcodes\n","      try:\n","        adata_hypr.var_names = var_names\n","      except:\n","        adata_hypr.var_names = var_names[1:]\n","\n","    if byprobe == False:\n","      df.columns = df.columns.str.split('_').str[0]\n","      collapsed_df = df.groupby(axis=1, level=0).sum()\n","      collapsed_X = collapsed_df.to_numpy()\n","      collapsed_var_names = collapsed_df.columns.to_numpy()\n","\n","      adata_hypr = anndata.AnnData(collapsed_X)\n","      adata_hypr.obs_names = barcodes\n","\n","      try:\n","        adata_hypr.var_names = collapsed_var_names\n","      except:\n","        adata_hypr.var_names = collapsed_var_names[1:]\n","\n","    return adata_hypr\n","\n","# Function for finding the intersecting barcodes between the modalities\n","def find_intersecting(mdata):\n","    \"\"\"\n","    Find intersecting barcodes and add 'intersecting' to obs with boolean values.\n","\n","    Parameters\n","    ----------\n","    mdata : object\n","        The mudata object.\n","\n","    Raises\n","    ------\n","    AssertionError\n","        If the length of mod_names is not equal to 2.\n","    \"\"\"\n","\n","    mod_names = list(mdata.mod.keys())\n","    assert len(mod_names) == 2, 'Function not implemented for mod_names with length different from 2'\n","\n","    obs_1, obs_2 = mdata[mod_names[0]].obs_names, mdata[mod_names[1]].obs_names\n","\n","    cmn_barcodes, idx_1, idx_2 = np.intersect1d(obs_1, obs_2, return_indices=True)\n","\n","    # Initializing 'intersecting' columns to False\n","    mdata[mod_names[0]].obs['intersecting'] = False\n","    mdata[mod_names[1]].obs['intersecting'] = False\n","\n","    # Updating 'intersecting' columns where True\n","    mdata[mod_names[0]].obs['intersecting'].iloc[idx_1] = True\n","    mdata[mod_names[1]].obs['intersecting'].iloc[idx_2] = True\n","\n","    logger.info(f\"Found {idx_1.size} intersecting barcodes\")\n","\n","\n","def annotate_genotype(mdata, variants, window=4, obs_key='mutant_type',\n","                      genotype_key='tapestri', phenotype_key='HyPR',\n","                      ignore_bystanders=False, write_all_as=False,\n","                      ignore_mixed=False, only_type=None):\n","\n","    adata = mdata.mod[genotype_key]\n","\n","    # loop through the variants in the variant list\n","    for variant in variants:\n","\n","        if obs_key in mdata.mod[phenotype_key].obs:\n","          pass\n","        else:\n","          mdata.mod[phenotype_key].obs[obs_key] = 'unannotated'\n","          mdata.mod[genotype_key].obs[obs_key] = 'unannotated'\n","\n","        try:\n","            idx = adata.var_names.get_loc(variant)\n","        except KeyError:\n","            logger.info(str(variant) + ' was not found in the genotype modality anndata')\n","            continue\n","\n","        # identify which cells have the variant\n","        bkg_idx, het_idx, hom_idx = [np.flatnonzero(adata.X[:, idx] == i) for i in (0, 1, 2)]\n","\n","        if only_type == 'het':\n","          hom_idx = None\n","        if only_type == 'hom':\n","          het_idx = None\n","\n","        if ignore_bystanders == True:\n","\n","          if het_idx is not None:\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[het_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","\n","            if write_all_as == False:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_het'\n","              mdata.mod[genotype_key].obs[obs_key][het_idx]= str(variant) + '_het'\n","\n","            else:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = write_all_as\n","\n","          if hom_idx is not None:\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[hom_idx],\n","                                                                   mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            if write_all_as == False:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype]= str(variant) + '_hom'\n","              mdata.mod[genotype_key].obs[obs_key][hom_idx]= str(variant) + '_hom'\n","            else:\n","              mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = write_all_as\n","\n","        else:\n","\n","          # identify indices of nearby SNPs that are within the bystander window\n","          chromosome, location = adata.varm['chrom'][idx], adata.varm['loc'][idx]\n","          nearby_ = np.where(adata.varm['chrom'] == chromosome)\n","          idx_ = np.where(np.logical_and(np.asarray(adata.varm['loc'], dtype=int) < int(location + window), np.asarray(adata.varm['loc'], dtype=int) > int(location - window)))\n","          nearby = np.intersect1d(nearby_, idx_)\n","\n","          mut_vals = adata.X[:, nearby]\n","          mut_vals[mut_vals == 3] = 0\n","          mut_vals[mut_vals > 1] = 1\n","          sum_muts = np.sum(mut_vals, axis=1)\n","\n","          if het_idx is not None:\n","\n","            pure_het_idx = np.intersect1d(np.argwhere(sum_muts == 1), het_idx)\n","            not_pure_het_idx = np.intersect1d(np.argwhere(sum_muts != 1), het_idx)\n","            mdata.mod[genotype_key].obs[obs_key][pure_het_idx] = str(variant) + '_het_pure'\n","            mdata.mod[genotype_key].obs[obs_key][not_pure_het_idx] = str(variant) + '_het_bystander'\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[not_pure_het_idx],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_het_bystander'\n","            logger.info(f\"Annotated {idx_genotype.size} heterozygous mutants with bystanders\")\n","\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[pure_het_idx],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_het_pure'\n","            logger.info(f\"Annotated {idx_genotype.size} pure heterozygous mutants\")\n","\n","          if hom_idx is not None:\n","\n","            pure_hom_idx = np.intersect1d(np.argwhere(sum_muts == 1), hom_idx)\n","            not_pure_hom_idx = np.intersect1d(np.argwhere(sum_muts != 1), hom_idx)\n","\n","            mdata.mod[genotype_key].obs[obs_key][pure_hom_idx] = str(variant) + '_hom_pure'\n","            mdata.mod[genotype_key].obs[obs_key][not_pure_hom_idx] = str(variant) + '_hom_bystander'\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[not_pure_hom_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_hom_bystander'\n","            logger.info(f\"Annotated {idx_genotype.size} homozygous mutants with bystanders\")\n","\n","            cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[pure_hom_idx],\n","                                                                    mdata.mod[phenotype_key].obs_names, return_indices=True)\n","            mdata.mod[phenotype_key].obs[obs_key][idx_phenotype] = str(variant) + '_hom_pure'\n","            logger.info(f\"Annotated {idx_genotype.size} pure homozygous mutants\")\n","\n","\n","    if ignore_mixed == False:\n","      # first, lets find the cells that have mixed edits of the intended variants\n","      indices = []\n","      for variant in variants:\n","          try:\n","              indices.append(adata.var_names.get_loc(variant))\n","          except KeyError:\n","              logger.info(str(variant) + ' was not found in the genotype modality anndata')\n","              continue\n","      idxs = np.asarray(indices).flatten()\n","\n","      edit_array = np.zeros(adata.X[:, idxs].shape)\n","      edit_array[adata.X[:, idxs] == 1] = 1\n","      edit_array[adata.X[:, idxs] == 2] = 1\n","      mixed_idxs = np.argwhere(np.sum(edit_array, axis=1) >= 2).ravel()\n","\n","      mdata.mod[genotype_key].obs[obs_key][mixed_idxs] = 'mixed_mutant'\n","      cmn_barcodes, idx_genotype, idx_phenotype = np.intersect1d(mdata.mod[genotype_key].obs_names[mixed_idxs],\n","                                                                      mdata.mod[phenotype_key].obs_names, return_indices=True)\n","      mdata.mod[phenotype_key].obs[obs_key][idx_phenotype]= 'mixed_mutant'"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"P0dAdH7x9D_8"},"outputs":[],"source":["import seaborn as sns\n","def barcode_rank_plot(adata, minimum=0, xmax=None):\n","\n","  # Sum the UMIs for each cell\n","  cell_umi_counts_all = adata.X.sum(axis=1)\n","  cell_idxs = np.argwhere(cell_umi_counts_all > minimum)\n","  cell_umi_counts = cell_umi_counts_all[cell_idxs]\n","\n","  # Convert to numpy array if it's not already\n","  cell_umi_counts = np.array(cell_umi_counts).flatten()\n","\n","  # Sort the UMI counts in descending order for the rank plot\n","  sorted_umi_counts = np.sort(cell_umi_counts)[::-1]\n","\n","  plt.figure(figsize=(5, 4), dpi=150)\n","  sns.lineplot(x=range(1, len(sorted_umi_counts) + 1), y=sorted_umi_counts)\n","  plt.xlabel('Barcode Rank')\n","  plt.ylabel('UMI Count')\n","  plt.title('Cell Barcode Rank Plot')\n","  plt.yscale('log')  # Log scale for better visualization\n","  if xmax is not None:\n","    plt.xlim(0, xmax)\n","  plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":20255,"status":"ok","timestamp":1712154695990,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"6Q4fkylJ9GvI","outputId":"2341138d-130d-4116-fc4a-d18f027a6623"},"outputs":[{"output_type":"stream","name":"stdout","text":["Mounted at /content/drive\n"]}],"source":["from google.colab import drive\n","drive.mount('/content/drive')"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"elapsed":1316,"status":"ok","timestamp":1712154701251,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"},"user_tz":240},"id":"Szn8RKcF9ICB","outputId":"9d306207-64b3-40ff-cd48-dce9c5fcabe9"},"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 6/6 [00:01<00:00,  5.14it/s]\n"]}],"source":["# define where all the files are on the drive\n","data_dir = '/path/to/tnrc18/data'\n","\n","file_locations = [['tapestri/HSC_Neu_D1_Day9.cells.loom', 'HyPR/Neu_D1_Day9_HyPR.txt'],\n","                  ['tapestri/HSC_Neu_D2_Day9.cells.loom', 'HyPR/Neu_D2_Day9_HyPR.txt'],\n","                  ['tapestri/HSC_Neu_D1_Day17.cells.loom', 'HyPR/Neu_D1_Day17_HyPR.txt'],\n","                  ['tapestri/HSC_Neu_D2_Day17.cells.loom', 'HyPR/Neu_D2_Day17_HyPR.txt']]\n","\n","sample_names = ['HSC_Neu_D1_Day9', 'HSC_Neu_D2_Day9', 'HSC_Neu_D1_Day17', 'HSC_Neu_D2_Day17']\n","\n","# check to make sure all those files exist\n","for i in tqdm(range(0, len(file_locations))):\n","  tap_file = data_dir + file_locations[i][0]\n","  hypr_file = data_dir + file_locations[i][1]\n","  if os.path.exists(tap_file):\n","      pass\n","  else:\n","    print(tap_file)\n","  if os.path.exists(hypr_file):\n","    pass\n","  else:\n","    print(hypr_file)"]},{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"eZoOqSeQ-9Zv","executionInfo":{"status":"ok","timestamp":1712154922869,"user_tz":240,"elapsed":209617,"user":{"displayName":"Anders Rasmussen","userId":"03752449904479783999"}},"outputId":"022971e6-ec97-4c0c-a1de-57b4a5c32591"},"outputs":[{"output_type":"stream","name":"stderr","text":["100%|██████████| 6/6 [03:29<00:00, 34.86s/it]\n"]}],"source":["mdata_list = []\n","for i in tqdm(range(0, len(file_locations))):\n","    adata_hypr = read_HyPR_file(data_dir + file_locations[i][1], byprobe=False)\n","    adata_tapestri = read_tapestri_loom(data_dir + file_locations[i][0])\n","    mdata_list.append(MuData({'HyPR': adata_hypr, 'tapestri': adata_tapestri}))"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"ha4M1wGp_DeQ"},"outputs":[],"source":["sns.set_theme()\n","for i, sample_name in enumerate(sample_names):\n","  print(sample_name)\n","  barcode_rank_plot(mdata_list[i].mod['HyPR'])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"PdUZGqXtAhn7"},"outputs":[],"source":["for i, sample_name in enumerate(sample_names):\n","  if i not in [2, 3]:\n","    scanpy.pp.filter_cells(mdata_list[i].mod['HyPR'], min_counts=1000)\n","  if i in [2, 3]:\n","    scanpy.pp.filter_cells(mdata_list[i].mod['HyPR'], min_counts=316)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"fawUQYnyBHcT"},"outputs":[],"source":["# find intersecting and remove any cells that are not intersecting across the modalities since we wont need them\n","for mdata in mdata_list:\n","  find_intersecting(mdata)\n","  mdata.mod['tapestri'] = mdata.mod['tapestri'][mdata.mod['tapestri'].obs['intersecting'] == True].copy()\n","  mdata.mod['HyPR'] = mdata.mod['HyPR'][mdata.mod['HyPR'].obs['intersecting'] == True].copy()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"VEhJeJgQGBqp"},"outputs":[],"source":["# identify the AAVS\n","chrom, loc, edit_type = [], [], []\n","variant_names = np.asarray(mdata.mod['tapestri'].var_names.values)\n","for i in range(0, np.size(variant_names)):\n","    chrom.append(variant_names[i].split(':')[0])\n","    loc.append(int(variant_names[i].split(':')[1]))\n","    edit_type.append(variant_names[i].split(':')[2])\n","\n","control_editing = []\n","for i in range(0, len(chrom)):\n","    if chrom[i] == 'chr19':\n","        if loc[i] in range(55115745, 55115764):\n","            if edit_type[i] == \"A/G\":\n","                control_editing.append(variant_names[i])\n","            if edit_type[i] == \"T/C\":\n","                control_editing.append(variant_names[i])\n","        if loc[i] in range(55115752, 55115771):\n","            if edit_type[i] == \"C/T\":\n","                control_editing.append(variant_names[i])\n","            if edit_type[i] == \"G/A\":\n","                control_editing.append(variant_names[i])"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"sT5B9q01GGoG"},"outputs":[],"source":["variants = ['chr7:5397122:C/T']\n","for i, mdata in enumerate(mdata_list):\n","    annotate_genotype(mdata, control_editing, ignore_bystanders=True, ignore_mixed=True, write_all_as='AAVS')\n","    annotate_genotype(mdata, variants, ignore_bystanders=False, ignore_mixed=False, window=10)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"LYUG-pCAHIP6"},"outputs":[],"source":["# create a concatenated anndata object for the HyPR\n","adata_hypr_list = [mdata.mod['HyPR'] for mdata in mdata_list]\n","adata_tapestri_list = [mdata.mod['tapestri'] for mdata in mdata_list]\n","\n","for anndata_obj, batch_name in zip(adata_hypr_list, sample_names):\n","    anndata_obj.obs['batch'] = batch_name\n","\n","for anndata_obj, batch_name in zip(adata_tapestri_list, sample_names):\n","    anndata_obj.obs['batch'] = batch_name\n","\n","concatenated_anndata = anndata.concat(adata_hypr_list)\n","concatenated_anndata.obs['S'] = concatenated_anndata.obs['batch'].str[-1].astype(int)\n","concatenated_anndata.obs['stimulation'] = concatenated_anndata.obs['batch'].str[:-1]"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"IKJYriudLgyJ"},"outputs":[],"source":["concatenated_anndata.layers['raw'] = concatenated_anndata.X.copy()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"pd_ZNeEZLigR"},"outputs":[],"source":["scanpy.pp.normalize_total(concatenated_anndata, target_sum=1e4)\n","scanpy.pp.log1p(concatenated_anndata)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"iD2PhPo6LnDx"},"outputs":[],"source":["scanpy.tl.pca(concatenated_anndata)\n","scanpy.pp.neighbors(concatenated_anndata)\n","scanpy.tl.umap(concatenated_anndata, min_dist=0.5, spread=3)"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"EkCrIXzoLn9n"},"outputs":[],"source":["# Check the UMAP\n","with plt.rc_context({\"figure.figsize\": (4, 3), \"figure.dpi\": (400)}):\n","  scanpy.pl.umap(concatenated_anndata, color='batch', show=False)\n","  ax = plt.gca()\n","  for spine in ax.spines.values():\n","      spine.set_visible(False)\n","  plt.show()"]},{"cell_type":"code","execution_count":null,"metadata":{"id":"waeCU8eTMbDk"},"outputs":[],"source":["# lets save this as an .h5ad for easy use in Python\n","output_path = \"where/to/write/h5ad\"\n","concatenated_anndata.write(output_path)"]}],"metadata":{"colab":{"machine_shape":"hm","provenance":[],"authorship_tag":"ABX9TyM0Y9k7nvvSk4kA4yejQsla"},"kernelspec":{"display_name":"Python 3","name":"python3"},"language_info":{"name":"python"}},"nbformat":4,"nbformat_minor":0}